// @tinkoff/linters/tslint/rules/tinkoffNewLineAroundControlStatementRule.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var Lint = require("tslint");
var ts = require("typescript");
var utils_1 = require("./utils");
var Rule = /** @class */ (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        var walker = new NewlineAroundControlStatementWalker(sourceFile, this.ruleName, undefined);
        return this.applyWithWalker(walker);
    };
    Rule.FAILURE_STRING = 'Missing blank line around control statement';
    Rule.metadata = {
        ruleName: 'tinkoff-new-line-around-control-statements',
        description: 'Enforces a new line around control statements',
        hasFix: true,
        optionsDescription: '',
        options: {},
        optionExamples: [],
        type: 'style',
        typescriptOnly: false,
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NewlineAroundControlStatementWalker = /** @class */ (function (_super) {
    __extends(NewlineAroundControlStatementWalker, _super);
    function NewlineAroundControlStatementWalker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NewlineAroundControlStatementWalker.prototype.getFailureString = function () {
        return Rule.FAILURE_STRING;
    };
    NewlineAroundControlStatementWalker.prototype.hasMatch = function (node) {
        switch (node.kind) {
            case ts.SyntaxKind.IfStatement:
            case ts.SyntaxKind.ForInStatement:
            case ts.SyntaxKind.ForOfStatement:
            case ts.SyntaxKind.ForStatement:
            case ts.SyntaxKind.WhileStatement:
            case ts.SyntaxKind.SwitchStatement:
            case ts.SyntaxKind.ReturnStatement:
                return true;
            default:
                return false;
        }
    };
    NewlineAroundControlStatementWalker.prototype.shouldSkipPrevious = function (node, previous) {
        return false;
    };
    return NewlineAroundControlStatementWalker;
}(utils_1.NewLineRuleWalker));
