// ng-bullet/bundles/ng-bullet.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core/testing')) :
    typeof define === 'function' && define.amd ? define('ng-bullet', ['exports', '@angular/core/testing'], factory) :
    (factory((global['ng-bullet'] = {}),global.ng.core.testing));
}(this, (function (exports,testing) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator.throw(value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [0, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    var _this = this;
    /**
     * Reconfigures current test suit to prevent angular components re-compilation after every test run.
     * Forces angular test bed to re-create zone and all injectable services by directly
     * setting _instantiated variable to false after every test run.
     * Cleanups all the changes and reverts test bed configuration after suite has finished.
     *
     * @param configureAction an optional delegate which can be used to configure test bed for the current test suite
     * directly in the configureTestSuite call (you don't need extra BeforeAll in this case)
     */
    var /** @type {?} */ configureTestSuite = function (configureAction) {
        var /** @type {?} */ testBedApi = testing.getTestBed();
        var /** @type {?} */ originReset = testing.TestBed.resetTestingModule;
        beforeAll(function () {
            testing.TestBed.resetTestingModule();
            testing.TestBed.resetTestingModule = function () { return testing.TestBed; };
        });
        if (configureAction) {
            beforeAll(function (done) {
                return (function () {
                    return __awaiter(_this, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0:
                                    configureAction();
                                    return [4 /*yield*/, testing.TestBed.compileComponents()];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    });
                })().then(done).catch(done.fail);
            });
        }
        afterEach(function () {
            testBedApi._activeFixtures.forEach(function (fixture) { return fixture.destroy(); });
            testBedApi._instantiated = false;
        });
        afterAll(function () {
            testing.TestBed.resetTestingModule = originReset;
            testing.TestBed.resetTestingModule();
        });
    };
    /**
     * A wrapper class around ComponentFixture, which provides useful accessros:
     * component - to access component instance of current the fixture
     * element - to access underlying native element of the current component
     * detectChanges - to run change detections using current fixture
     * resolve - to resolve a type using current fixture's injector
     */
    var /**
     * A wrapper class around ComponentFixture, which provides useful accessros:
     * component - to access component instance of current the fixture
     * element - to access underlying native element of the current component
     * detectChanges - to run change detections using current fixture
     * resolve - to resolve a type using current fixture's injector
     */ TestCtx = (function () {
        function TestCtx(fixture) {
            this.fixture = fixture;
        }
        Object.defineProperty(TestCtx.prototype, "component", {
            get: /**
             * @return {?}
             */ function () { return this.fixture.componentInstance; },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TestCtx.prototype, "element", {
            get: /**
             * @return {?}
             */ function () { return this.fixture.debugElement.nativeElement; },
            enumerable: true,
            configurable: true
        });
        /**
         * @return {?}
         */
        TestCtx.prototype.detectChanges = /**
         * @return {?}
         */
            function () { this.fixture.detectChanges(); };
        /**
         * @param {?} component
         * @return {?}
         */
        TestCtx.prototype.resolve = /**
         * @param {?} component
         * @return {?}
         */
            function (component) { return this.fixture.debugElement.injector.get(component); };
        return TestCtx;
    }());
    /**
     * Creates TestCtx instance for the Angular Component which is not initialized yet (no ngOnInit called)
     * Use case: you can override Component's providers before hooks are called.
     *
     * @param component - type of component to create instance of
     * *
     */
    var /** @type {?} */ createTestContext = function (component) {
        var /** @type {?} */ fixture = testing.TestBed.createComponent(component);
        var /** @type {?} */ testCtx = new TestCtx(fixture);
        return testCtx;
    };
    /**
     * Same as \@function createTestContext, but waits till fixture becomes stable
     */
    var /** @type {?} */ createStableTestContext = function (component) {
        return __awaiter(_this, void 0, void 0, function () {
            var testCtx;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        testCtx = createTestContext(component);
                        testCtx.detectChanges();
                        return [4 /*yield*/, testCtx.fixture.whenStable()];
                    case 1:
                        _a.sent();
                        testCtx.detectChanges();
                        return [2 /*return*/, testCtx];
                }
            });
        });
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.configureTestSuite = configureTestSuite;
    exports.TestCtx = TestCtx;
    exports.createTestContext = createTestContext;
    exports.createStableTestContext = createStableTestContext;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
