// @ng-web-apis/universal/esm2015/classes/storage-mock.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class StorageMock {
    constructor() {
        this.storage = new Map();
    }
    /**
     * @return {?}
     */
    get length() {
        return this.storage.size;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    getItem(key) {
        return this.storage.has(key) ? (/** @type {?} */ (this.storage.get(key))) : null;
    }
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    setItem(key, value) {
        this.storage.set(key, value);
    }
    /**
     * @return {?}
     */
    clear() {
        this.storage.clear();
    }
    /**
     * @param {?} index
     * @return {?}
     */
    key(index) {
        return index < this.storage.size ? [...this.storage.keys()][index] : null;
    }
    /**
     * @param {?} key
     * @return {?}
     */
    removeItem(key) {
        this.storage.delete(key);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    StorageMock.prototype.storage;
}
