// @ng-web-apis/universal/esm5/utils/provide-location.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { SSR_LOCATION } from '../tokens/ssr-location';
import { empty } from './functions';
/**
 * @param {?} req
 * @return {?}
 */
export function provideLocation(req) {
    /** @type {?} */
    var protocol = 'encrypted' in req.socket ? 'https' : 'http';
    /** @type {?} */
    var url = new URL(protocol + "://" + req.headers['host'] + req.url);
    url.assign = empty;
    url.reload = empty;
    url.replace = empty;
    url.ancestorOrigins = new (/** @class */ (function (_super) {
        tslib_1.__extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @return {?}
         */
        class_1.prototype.contains = /**
         * @return {?}
         */
        function () {
            return false;
        };
        /**
         * @return {?}
         */
        class_1.prototype.item = /**
         * @return {?}
         */
        function () {
            return null;
        };
        return class_1;
    }(Array)))();
    return {
        provide: SSR_LOCATION,
        useValue: url,
    };
}
