// @angular/cdk/esm2015/table/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: src/cdk/table/public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
export { DataRowOutlet, HeaderRowOutlet, FooterRowOutlet, CDK_TABLE_TEMPLATE, CdkTable } from './table';
export { CdkCellDef, CdkHeaderCellDef, CdkFooterCellDef, CdkColumnDef, BaseCdkCell, CdkHeaderCell, CdkFooterCell, CdkCell } from './cell';
export { CDK_ROW_TEMPLATE, BaseRowDef, CdkHeaderRowDef, CdkFooterRowDef, CdkRowDef, CdkCellOutlet, CdkHeaderRow, CdkFooterRow, CdkRow } from './row';
export { CdkTableModule } from './table-module';
export { STICKY_DIRECTIONS, StickyStyler } from './sticky-styler';
export { mixinHasStickyInput } from './can-stick';
export { CdkTextColumn } from './text-column';
export { CDK_TABLE, TEXT_COLUMN_OPTIONS } from './tokens';
/** Re-export DataSource for a more intuitive experience for users of just the table. */
export { DataSource } from '@angular/cdk/collections';
