// @angular/cli/commands/analytics.js
"use strict";
// THIS FILE IS AUTOMATICALLY GENERATED. TO UPDATE THIS FILE YOU NEED TO CHANGE THE
// CORRESPONDING JSON SCHEMA FILE, THEN RUN devkit-admin build (or bazel build ...).
Object.defineProperty(exports, "__esModule", { value: true });
var HelpEnum;
(function (HelpEnum) {
    HelpEnum["HelpJson"] = "JSON";
    HelpEnum["Json"] = "json";
})(HelpEnum = exports.HelpEnum || (exports.HelpEnum = {}));
/**
 * Sets the default analytics enablement status for the project.
 */
var ProjectSetting;
(function (ProjectSetting) {
    ProjectSetting["Off"] = "off";
    ProjectSetting["On"] = "on";
    ProjectSetting["Prompt"] = "prompt";
})(ProjectSetting = exports.ProjectSetting || (exports.ProjectSetting = {}));
/**
 * Directly enables or disables all usage analytics for the user, or prompts the user to set
 * the status interactively, or sets the default status for the project.
 */
var SettingOrProject;
(function (SettingOrProject) {
    SettingOrProject["Ci"] = "ci";
    SettingOrProject["Off"] = "off";
    SettingOrProject["On"] = "on";
    SettingOrProject["Project"] = "project";
    SettingOrProject["Prompt"] = "prompt";
})(SettingOrProject = exports.SettingOrProject || (exports.SettingOrProject = {}));
