// @angular/cdk/esm5/overlay/overlay-config.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __values } from "tslib";
import { NoopScrollStrategy } from './scroll/index';
/** Initial configuration used when creating an overlay. */
var OverlayConfig = /** @class */ (function () {
    function OverlayConfig(config) {
        var e_1, _a;
        /** Strategy to be used when handling scroll events while the overlay is open. */
        this.scrollStrategy = new NoopScrollStrategy();
        /** Custom class to add to the overlay pane. */
        this.panelClass = '';
        /** Whether the overlay has a backdrop. */
        this.hasBackdrop = false;
        /** Custom class to add to the backdrop */
        this.backdropClass = 'cdk-overlay-dark-backdrop';
        /**
         * Whether the overlay should be disposed of when the user goes backwards/forwards in history.
         * Note that this usually doesn't include clicking on links (unless the user is using
         * the `HashLocationStrategy`).
         */
        this.disposeOnNavigation = false;
        if (config) {
            // Use `Iterable` instead of `Array` because TypeScript, as of 3.6.3,
            // loses the array generic type in the `for of`. But we *also* have to use `Array` because
            // typescript won't iterate over an `Iterable` unless you compile with `--downlevelIteration`
            var configKeys = Object.keys(config);
            try {
                for (var configKeys_1 = __values(configKeys), configKeys_1_1 = configKeys_1.next(); !configKeys_1_1.done; configKeys_1_1 = configKeys_1.next()) {
                    var key = configKeys_1_1.value;
                    if (config[key] !== undefined) {
                        // TypeScript, as of version 3.5, sees the left-hand-side of this expression
                        // as "I don't know *which* key this is, so the only valid value is the intersection
                        // of all the posible values." In this case, that happens to be `undefined`. TypeScript
                        // is not smart enough to see that the right-hand-side is actually an access of the same
                        // exact type with the same exact key, meaning that the value type must be identical.
                        // So we use `any` to work around this.
                        this[key] = config[key];
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (configKeys_1_1 && !configKeys_1_1.done && (_a = configKeys_1.return)) _a.call(configKeys_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
    }
    return OverlayConfig;
}());
export { OverlayConfig };
