// @ng-web-apis/universal/fesm5/ng-web-apis-universal.js
import { NEVER } from 'rxjs';
import { InjectionToken, Optional } from '@angular/core';
import { ANIMATION_FRAME, LOCAL_STORAGE, USER_AGENT, NAVIGATOR, PERFORMANCE, SESSION_STORAGE } from '@ng-web-apis/common';
import { __spread, __extends, __assign } from 'tslib';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var UNIVERSAL_ANIMATION_FRAME = {
    provide: ANIMATION_FRAME,
    useValue: NEVER,
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var StorageMock = /** @class */ (function () {
    function StorageMock() {
        this.storage = new Map();
    }
    Object.defineProperty(StorageMock.prototype, "length", {
        get: /**
         * @return {?}
         */
        function () {
            return this.storage.size;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} key
     * @return {?}
     */
    StorageMock.prototype.getItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return this.storage.has(key) ? (/** @type {?} */ (this.storage.get(key))) : null;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    StorageMock.prototype.setItem = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        this.storage.set(key, value);
    };
    /**
     * @return {?}
     */
    StorageMock.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.storage.clear();
    };
    /**
     * @param {?} index
     * @return {?}
     */
    StorageMock.prototype.key = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        return index < this.storage.size ? __spread(this.storage.keys())[index] : null;
    };
    /**
     * @param {?} key
     * @return {?}
     */
    StorageMock.prototype.removeItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        this.storage.delete(key);
    };
    return StorageMock;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var UNIVERSAL_LOCAL_STORAGE = {
    provide: LOCAL_STORAGE,
    useClass: StorageMock,
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var SSR_LOCATION = new InjectionToken('Location object passed from server side');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @template T
 * @param {?} v
 * @return {?}
 */
function identity(v) {
    return v;
}
/**
 * @return {?}
 */
function empty() { }
/**
 * @return {?}
 */
function emptyArray() {
    return [];
}
/**
 * @return {?}
 */
function emptyObject() {
    return {};
}
/**
 * @return {?}
 */
function alwaysFalse() {
    return false;
}
/**
 * @return {?}
 */
function alwaysZero() {
    return 0;
}
/**
 * @return {?}
 */
function alwaysRejected() {
    return Promise.reject().catch(empty);
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var UNIVERSAL_LOCATION = {
    provide: USER_AGENT,
    deps: [SSR_LOCATION],
    useFactory: identity,
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var SSR_USER_AGENT = new InjectionToken('User Agent string passed from server side');

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var EVENT_TARGET = {
    addEventListener: empty,
    dispatchEvent: alwaysFalse,
    removeEventListener: empty,
};
/** @type {?} */
var PLUGIN = new (/** @class */ (function (_super) {
    __extends(class_1, _super);
    function class_1() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.description = '';
        _this.filename = '';
        _this.name = '';
        _this.version = '';
        return _this;
    }
    /**
     * @return {?}
     */
    class_1.prototype.item = /**
     * @return {?}
     */
    function () {
        return {
            description: '',
            enabledPlugin: this,
            suffixes: '',
            type: '',
        };
    };
    /**
     * @return {?}
     */
    class_1.prototype.namedItem = /**
     * @return {?}
     */
    function () {
        return {
            description: '',
            enabledPlugin: this,
            suffixes: '',
            type: '',
        };
    };
    return class_1;
}(Array)))();
/** @type {?} */
var MIME_TYPES = new (/** @class */ (function (_super) {
    __extends(class_2, _super);
    function class_2() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    class_2.prototype.item = /**
     * @return {?}
     */
    function () {
        return PLUGIN;
    };
    /**
     * @return {?}
     */
    class_2.prototype.namedItem = /**
     * @return {?}
     */
    function () {
        return PLUGIN;
    };
    /**
     * @return {?}
     */
    class_2.prototype.refresh = /**
     * @return {?}
     */
    function () { };
    return class_2;
}(Array)))();
/**
 * @param {?} userAgent
 * @return {?}
 */
function navigatorFactory(userAgent) {
    return {
        appCodeName: '',
        appName: '',
        appVersion: '',
        platform: '',
        product: '',
        productSub: '',
        userAgent: userAgent || '',
        vendor: '',
        vendorSub: '',
        onLine: false,
        confirmSiteSpecificTrackingException: alwaysFalse,
        confirmWebWideTrackingException: alwaysFalse,
        removeSiteSpecificTrackingException: empty,
        removeWebWideTrackingException: empty,
        storeSiteSpecificTrackingException: empty,
        storeWebWideTrackingException: empty,
        msSaveBlob: alwaysFalse,
        msSaveOrOpenBlob: alwaysFalse,
        sendBeacon: alwaysFalse,
        hardwareConcurrency: 0,
        getDisplayMedia: alwaysRejected,
        language: '',
        languages: [],
        storage: {
            estimate: alwaysRejected,
            persist: alwaysRejected,
            persisted: alwaysRejected,
        },
        activeVRDisplays: [],
        authentication: {
            getAssertion: alwaysRejected,
            makeCredential: alwaysRejected,
        },
        clipboard: __assign({}, EVENT_TARGET, { readText: alwaysRejected, writeText: alwaysRejected }),
        cookieEnabled: false,
        doNotTrack: null,
        gamepadInputEmulation: 'keyboard',
        geolocation: {
            clearWatch: empty,
            getCurrentPosition: empty,
            watchPosition: alwaysZero,
        },
        maxTouchPoints: 0,
        mediaDevices: __assign({}, EVENT_TARGET, { ondevicechange: null, enumerateDevices: alwaysRejected, getSupportedConstraints: emptyObject, getUserMedia: alwaysRejected }),
        mimeTypes: MIME_TYPES,
        msManipulationViewsEnabled: false,
        msMaxTouchPoints: 0,
        msPointerEnabled: false,
        permissions: {
            query: alwaysRejected,
        },
        plugins: MIME_TYPES,
        pointerEnabled: false,
        serviceWorker: __assign({}, EVENT_TARGET, { controller: null, oncontrollerchange: null, onmessage: null, onmessageerror: null, ready: alwaysRejected(), getRegistration: alwaysRejected, getRegistrations: alwaysRejected, register: alwaysRejected, startMessages: empty }),
        webdriver: false,
        getGamepads: emptyArray,
        getUserMedia: empty,
        getVRDisplays: alwaysRejected,
        javaEnabled: alwaysFalse,
        msLaunchUri: empty,
        requestMediaKeySystemAccess: alwaysRejected,
        vibrate: alwaysFalse,
    };
}
/** @type {?} */
var UNIVERSAL_NAVIGATOR = {
    provide: NAVIGATOR,
    deps: [[new Optional(), SSR_USER_AGENT]],
    useFactory: navigatorFactory,
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @return {?}
 */
function performanceFactory() {
    return require('perf_hooks').performance;
}
/** @type {?} */
var UNIVERSAL_PERFORMANCE = {
    provide: PERFORMANCE,
    deps: [],
    useFactory: performanceFactory,
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var UNIVERSAL_SESSION_STORAGE = {
    provide: SESSION_STORAGE,
    useClass: StorageMock,
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var MOCK = {
    paused: false,
    pending: false,
    speaking: false,
    onvoiceschanged: empty,
    addEventListener: empty,
    removeEventListener: empty,
    dispatchEvent: alwaysFalse,
    cancel: empty,
    pause: empty,
    resume: empty,
    speak: empty,
    getVoices: emptyArray,
};
/** @type {?} */
var UNIVERSAL_SPEECH_SYNTHESIS = {
    provide: LOCAL_STORAGE,
    useValue: MOCK,
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var UNIVERSAL_USER_AGENT = {
    provide: USER_AGENT,
    deps: [SSR_USER_AGENT],
    useFactory: identity,
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var UNIVERSAL_PROVIDERS = [
    UNIVERSAL_ANIMATION_FRAME,
    UNIVERSAL_LOCAL_STORAGE,
    UNIVERSAL_SESSION_STORAGE,
    UNIVERSAL_LOCATION,
    UNIVERSAL_NAVIGATOR,
    UNIVERSAL_PERFORMANCE,
    UNIVERSAL_SPEECH_SYNTHESIS,
    UNIVERSAL_USER_AGENT,
];
/**
 * @deprecated use {\@link UNIVERSAL_PROVIDERS}
 * @type {?}
 */
var UNIVERSAL_TOKENS = UNIVERSAL_PROVIDERS;

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} req
 * @return {?}
 */
function provideLocation(req) {
    /** @type {?} */
    var protocol = 'encrypted' in req.socket ? 'https' : 'http';
    /** @type {?} */
    var url = new URL(protocol + "://" + req.headers['host'] + req.url);
    url.assign = empty;
    url.reload = empty;
    url.replace = empty;
    url.ancestorOrigins = new (/** @class */ (function (_super) {
        __extends(class_1, _super);
        function class_1() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @return {?}
         */
        class_1.prototype.contains = /**
         * @return {?}
         */
        function () {
            return false;
        };
        /**
         * @return {?}
         */
        class_1.prototype.item = /**
         * @return {?}
         */
        function () {
            return null;
        };
        return class_1;
    }(Array)))();
    return {
        provide: SSR_LOCATION,
        useValue: url,
    };
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} req
 * @return {?}
 */
function provideUserAgent(req) {
    return {
        provide: SSR_USER_AGENT,
        useValue: req.headers['user-agent'],
    };
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { UNIVERSAL_ANIMATION_FRAME, UNIVERSAL_LOCAL_STORAGE, UNIVERSAL_LOCATION, navigatorFactory, UNIVERSAL_NAVIGATOR, performanceFactory, UNIVERSAL_PERFORMANCE, UNIVERSAL_SESSION_STORAGE, UNIVERSAL_SPEECH_SYNTHESIS, UNIVERSAL_PROVIDERS, UNIVERSAL_TOKENS, UNIVERSAL_USER_AGENT, provideLocation, provideUserAgent, StorageMock as ɵa, SSR_LOCATION as ɵb, SSR_USER_AGENT as ɵj, alwaysFalse as ɵg, alwaysRejected as ɵi, alwaysZero as ɵh, empty as ɵd, emptyArray as ɵe, emptyObject as ɵf, identity as ɵc };
