// ng-bullet/fesm5/ng-bullet.js
import { __awaiter, __generator } from 'tslib';
import '@angular/core';
import { getTestBed, TestBed } from '@angular/core/testing';

var _this = this;
/**
 * Reconfigures current test suit to prevent angular components re-compilation after every test run.
 * Forces angular test bed to re-create zone and all injectable services by directly
 * setting _instantiated variable to false after every test run.
 * Cleanups all the changes and reverts test bed configuration after suite has finished.
 *
 * @param configureAction an optional delegate which can be used to configure test bed for the current test suite
 * directly in the configureTestSuite call (you don't need extra BeforeAll in this case)
 */
var /** @type {?} */ configureTestSuite = function (configureAction) {
    var /** @type {?} */ testBedApi = getTestBed();
    var /** @type {?} */ originReset = TestBed.resetTestingModule;
    beforeAll(function () {
        TestBed.resetTestingModule();
        TestBed.resetTestingModule = function () { return TestBed; };
    });
    if (configureAction) {
        beforeAll(function (done) {
            return (function () { return __awaiter(_this, void 0, void 0, function () {
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            configureAction();
                            return [4 /*yield*/, TestBed.compileComponents()];
                        case 1:
                            _a.sent();
                            return [2 /*return*/];
                    }
                });
            }); })().then(done).catch(done.fail);
        });
    }
    afterEach(function () {
        testBedApi._activeFixtures.forEach(function (fixture) { return fixture.destroy(); });
        testBedApi._instantiated = false;
    });
    afterAll(function () {
        TestBed.resetTestingModule = originReset;
        TestBed.resetTestingModule();
    });
};
/**
 * A wrapper class around ComponentFixture, which provides useful accessros:
 * component - to access component instance of current the fixture
 * element - to access underlying native element of the current component
 * detectChanges - to run change detections using current fixture
 * resolve - to resolve a type using current fixture's injector
 */
var  /**
 * A wrapper class around ComponentFixture, which provides useful accessros:
 * component - to access component instance of current the fixture
 * element - to access underlying native element of the current component
 * detectChanges - to run change detections using current fixture
 * resolve - to resolve a type using current fixture's injector
 */
TestCtx = /** @class */ (function () {
    function TestCtx(fixture) {
        this.fixture = fixture;
    }
    Object.defineProperty(TestCtx.prototype, "component", {
        get: /**
         * @return {?}
         */
        function () { return this.fixture.componentInstance; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TestCtx.prototype, "element", {
        get: /**
         * @return {?}
         */
        function () { return this.fixture.debugElement.nativeElement; },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TestCtx.prototype.detectChanges = /**
     * @return {?}
     */
    function () { this.fixture.detectChanges(); };
    /**
     * @param {?} component
     * @return {?}
     */
    TestCtx.prototype.resolve = /**
     * @param {?} component
     * @return {?}
     */
    function (component) { return this.fixture.debugElement.injector.get(component); };
    return TestCtx;
}());
/**
 * Creates TestCtx instance for the Angular Component which is not initialized yet (no ngOnInit called)
 * Use case: you can override Component's providers before hooks are called.
 *
 * @param component - type of component to create instance of
 * *
 */
var /** @type {?} */ createTestContext = function (component) {
    var /** @type {?} */ fixture = TestBed.createComponent(component);
    var /** @type {?} */ testCtx = new TestCtx(fixture);
    return testCtx;
};
/**
 * Same as \@function createTestContext, but waits till fixture becomes stable
 */
var /** @type {?} */ createStableTestContext = function (component) { return __awaiter(_this, void 0, void 0, function () {
    var testCtx;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                testCtx = createTestContext(component);
                testCtx.detectChanges();
                return [4 /*yield*/, testCtx.fixture.whenStable()];
            case 1:
                _a.sent();
                testCtx.detectChanges();
                return [2 /*return*/, testCtx];
        }
    });
}); };

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { configureTestSuite, TestCtx, createTestContext, createStableTestContext };
