// @angular/animations/esm2015/browser/src/dsl/animation_dsl_visitor.js
/**
 * @fileoverview added by tsickle
 * Generated from: packages/animations/browser/src/dsl/animation_dsl_visitor.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function AnimationDslVisitor() { }
if (false) {
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitTrigger = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitState = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitTransition = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitSequence = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitGroup = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitAnimate = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitStyle = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitKeyframes = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitReference = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitAnimateChild = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitAnimateRef = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitQuery = function (node, context) { };
    /**
     * @param {?} node
     * @param {?} context
     * @return {?}
     */
    AnimationDslVisitor.prototype.visitStagger = function (node, context) { };
}
