// tslint-teamcity-reporter/lib/Reporter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Formatter = void 0;

var _stringify = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/json/stringify"));

var _inheritsLoose2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/inheritsLoose"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));

var _abstractFormatter = require("tslint/lib/language/formatter/abstractFormatter");

var _util = require("./util");

var _inspections = require("./formatters/inspections");

var _errors = require("./formatters/errors");

var Formatter =
/*#__PURE__*/
function (_AbstractFormatter) {
  (0, _inheritsLoose2.default)(Formatter, _AbstractFormatter);

  function Formatter() {
    return _AbstractFormatter.apply(this, arguments) || this;
  }

  var _proto = Formatter.prototype;

  /* tslint:disable:object-literal-sort-keys */

  /* tslint:enable:object-literal-sort-keys */
  _proto.format = function format(failures, config) {
    if (config === void 0) {
      config = {};
    }

    var userConfig = (0, _util.getUserConfig)(config);

    if (process.env.TSLINT_TEAMCITY_DISPLAY_CONFIG) {
      // tslint:disable-next-line no-console
      console.info("Running TSLint Teamcity with config: " + (0, _stringify.default)(userConfig, null, 4));
    }

    var outputMessage = '';

    switch (userConfig.reporter.toLowerCase()) {
      case 'inspections':
        outputMessage = (0, _inspections.formatAsInspections)(failures, userConfig);
        break;

      case 'errors':
      default:
        outputMessage = (0, _errors.formatAsTests)(failures, userConfig);
        break;
    }

    return outputMessage;
  };

  return Formatter;
}(_abstractFormatter.AbstractFormatter);

exports.Formatter = Formatter;
(0, _defineProperty2.default)(Formatter, "metadata", {
  formatterName: 'json',
  description: 'Formats errors as TeamCity report.',
  sample: "\n##teamcity[testSuiteStarted name='TSLint Violations']\n##teamcity[testStarted name='TSLint Violations: test1.ts']\n##teamcity[testFailed name='TSLint Violations: test1.ts' message='line 0, col 0, first failure (first-name)|nline 8, col 1, last failure (last-name)']\n##teamcity[testStdOut name='TSLint Violations: test1.ts' out='warning: line 0, col 0, full failure (full-name)']\n##teamcity[testFinished name='TSLint Violations: test1.ts']\n##teamcity[testStarted name='TSLint Violations: test2.ts']\n##teamcity[testStdOut name='TSLint Violations: test2.ts' out='warning: line 8, col 1, full failure (full-name)']\n##teamcity[testFinished name='TSLint Violations: test2.ts']\n##teamcity[testSuiteFinished name='TSLint Violations']\n##teamcity[buildStatisticValue key='TSLint Error Count' value='2']\n##teamcity[buildStatisticValue key='TSLint Warning Count' value='2']".slice(1),
  consumer: 'machine'
});
