// @ng-web-apis/mutation-observer/esm5/utils/mutation-observer-init-factory.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { booleanAttribute } from './boolean-attribute';
/**
 * @param {?} __0
 * @return {?}
 */
export function mutationObserverInitFactory(_a) {
    var nativeElement = _a.nativeElement;
    /** @type {?} */
    var attributeFilterAttr = nativeElement.getAttribute('attributeFilter');
    /** @type {?} */
    var attributeFilter = attributeFilterAttr
        ? attributeFilterAttr.split(',').map(function (attr) { return attr.trim(); })
        : undefined;
    return {
        attributeFilter: attributeFilter,
        attributeOldValue: booleanAttribute(nativeElement, 'attributeOldValue'),
        attributes: booleanAttribute(nativeElement, 'attributes'),
        characterData: booleanAttribute(nativeElement, 'characterData'),
        characterDataOldValue: booleanAttribute(nativeElement, 'characterDataOldValue'),
        childList: booleanAttribute(nativeElement, 'childList'),
        subtree: booleanAttribute(nativeElement, 'subtree'),
    };
}
