// @tinkoff/ng-dompurify/esm2015/lib/ng-dompurify.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DOCUMENT } from '@angular/common';
import { Inject, Injectable, SecurityContext } from '@angular/core';
import * as dompurify from 'dompurify';
import { DOMPURIFY_CONFIG } from './tokens/dompurify-config';
import { DOMPURIFY_HOOKS } from './tokens/dompurify-hooks';
import { SANITIZE_STYLE } from './tokens/sanitize-style';
import * as i0 from "@angular/core";
import * as i1 from "./tokens/dompurify-config";
import * as i2 from "./tokens/sanitize-style";
import * as i3 from "@angular/common";
import * as i4 from "./tokens/dompurify-hooks";
/** @type {?} */
const createDOMPurify = dompurify;
/**
 * Implementation of Angular {\@link Sanitizer} purifying via DOMPurify
 *
 * use {\@link DOMPURIFY_CONFIG} token to provide config ({\@link NgDompurifyConfig})
 * use {\@link SANITIZE_STYLE} token to provide a style sanitizing method ({\@link SanitizeStyle})
 * use {\@link DOMPURIFY_HOOKS} token to provide a hooks for DOMPurify ({\@link addHook})
 *
 * Ambient type cannot be used without \@dynamic https://github.com/angular/angular/issues/23395
 * \@dynamic
 */
export class NgDompurifySanitizer {
    /**
     * @param {?} config
     * @param {?} sanitizeStyle
     * @param {?} __2
     * @param {?} hooks
     */
    constructor(config, sanitizeStyle, { defaultView }, hooks) {
        this.config = config;
        this.sanitizeStyle = sanitizeStyle;
        this.domPurify = createDOMPurify((/** @type {?} */ (defaultView)));
        hooks.forEach(({ name, hook }) => {
            this.domPurify.addHook(name, hook);
        });
    }
    /**
     * @param {?} context
     * @param {?} value
     * @param {?=} config
     * @return {?}
     */
    sanitize(context, value, config = this.config) {
        if (context === SecurityContext.SCRIPT) {
            throw new Error('DOMPurify does not support SCRIPT context');
        }
        return context === SecurityContext.STYLE
            ? this.sanitizeStyle(String(value))
            : this.domPurify.sanitize(String(value || ''), config);
    }
}
NgDompurifySanitizer.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/** @nocollapse */
NgDompurifySanitizer.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOMPURIFY_CONFIG,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [SANITIZE_STYLE,] }] },
    { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [DOMPURIFY_HOOKS,] }] }
];
/** @nocollapse */ NgDompurifySanitizer.ngInjectableDef = i0.defineInjectable({ factory: function NgDompurifySanitizer_Factory() { return new NgDompurifySanitizer(i0.inject(i1.DOMPURIFY_CONFIG), i0.inject(i2.SANITIZE_STYLE), i0.inject(i3.DOCUMENT), i0.inject(i4.DOMPURIFY_HOOKS)); }, token: NgDompurifySanitizer, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgDompurifySanitizer.prototype.domPurify;
    /**
     * @type {?}
     * @private
     */
    NgDompurifySanitizer.prototype.config;
    /**
     * @type {?}
     * @private
     */
    NgDompurifySanitizer.prototype.sanitizeStyle;
}
