// @ng-web-apis/intersection-observer/esm5/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Public API Surface of @ng-web-apis/intersection-observer
 */
/* Directives */
export { IntersectionObserveeDirective } from './directives/intersection-observee.directive';
export { IntersectionObserverDirective } from './directives/intersection-observer.directive';
export { IntersectionRootDirective } from './directives/intersection-root.directive';
/* Modules */
export { IntersectionObserverModule } from './module';
/* Services */
export { IntersectionObserveeService } from './services/intersection-observee.service';
export { IntersectionObserverService } from './services/intersection-observer.service';
/* Tokens */
export { INTERSECTION_ROOT } from './tokens/intersection-root';
export { INTERSECTION_ROOT_MARGIN_DEFAULT, INTERSECTION_ROOT_MARGIN } from './tokens/intersection-root-margin';
export { INTERSECTION_THRESHOLD_DEFAULT, INTERSECTION_THRESHOLD } from './tokens/intersection-threshold';
export { INTERSECTION_OBSERVER_SUPPORT } from './tokens/support';
