// @ng-web-apis/resize-observer/esm2015/directives/resize-observer.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Attribute, Directive, ElementRef, Inject, Output } from '@angular/core';
import { Observable } from 'rxjs';
import { ResizeObserverService } from '../services/resize-observer.service';
import { RESIZE_OPTION_BOX, RESIZE_OPTION_BOX_DEFAULT } from '../tokens/resize-option-box';
// TODO switch to Attribute once https://github.com/angular/angular/issues/36479 is fixed
/**
 * @param {?} __0
 * @return {?}
 */
export function boxExtractor({ nativeElement, }) {
    /** @type {?} */
    const attribute = (/** @type {?} */ (nativeElement.getAttribute('waResizeBox')));
    return boxFactory(attribute);
}
/**
 * @param {?} box
 * @return {?}
 */
export function boxFactory(box) {
    return box || RESIZE_OPTION_BOX_DEFAULT;
}
// @dynamic
export class ResizeObserverDirective {
    /**
     * @param {?} entries$
     * @param {?} _box
     */
    constructor(entries$, _box) {
        this.waResizeObserver = entries$;
    }
}
ResizeObserverDirective.decorators = [
    { type: Directive, args: [{
                selector: '[waResizeObserver]',
                providers: [
                    ResizeObserverService,
                    {
                        provide: RESIZE_OPTION_BOX,
                        deps: [ElementRef],
                        useFactory: boxExtractor,
                    },
                ],
            },] }
];
/** @nocollapse */
ResizeObserverDirective.ctorParameters = () => [
    { type: Observable, decorators: [{ type: Inject, args: [ResizeObserverService,] }] },
    { type: undefined, decorators: [{ type: Attribute, args: ['waResizeBox',] }] }
];
ResizeObserverDirective.propDecorators = {
    waResizeObserver: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    ResizeObserverDirective.prototype.waResizeObserver;
}
