// @tinkoff/ng-polymorpheus/esm2015/outlet/outlet.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgComponentOutlet } from '@angular/common';
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ContentChild, Inject, Injector, Input, TemplateRef, ViewChild, } from '@angular/core';
import { PolymorpheusComponent } from '../classes/component';
import { PolymorpheusTemplate } from '../directives/template';
/**
 * Outlet instantiating {\@link PolymorpheusContent} with given context
 * @template C
 */
export class PolymorpheusOutletComponent {
    /**
     * @param {?} injector
     */
    constructor(injector) {
        this.injector = injector;
        this.content = null;
        this.template = null;
    }
    /**
     * @return {?}
     */
    get primitive() {
        if (!this.content ||
            this.isComponent(this.content) ||
            this.isTemplate(this.content)) {
            return '';
        }
        return typeof this.content === 'function'
            ? this.content(this.context)
            : this.content;
    }
    /**
     * @param {?} content
     * @return {?}
     */
    isDirective(content) {
        return content instanceof PolymorpheusTemplate;
    }
    /**
     * @param {?} content
     * @return {?}
     */
    isTemplate(content) {
        return this.isDirective(content) || content instanceof TemplateRef;
    }
    /**
     * @param {?} content
     * @return {?}
     */
    isComponent(content) {
        return content instanceof PolymorpheusComponent;
    }
    /**
     * @param {?} content
     * @param {?} componentTmp
     * @return {?}
     */
    getTemplate(content, componentTmp) {
        if (this.isComponent(content)) {
            return componentTmp;
        }
        return this.isDirective(content) ? content.template : content;
    }
    /**
     * @param {?} __0
     * @return {?}
     */
    ngOnChanges({ content, context }) {
        // TODO: Keep an eye on private field, name can change
        /** @type {?} */
        const componentRef = (/** @type {?} */ ((!content &&
            context &&
            this.outlet &&
            this.outlet['_componentRef'])));
        if (componentRef) {
            componentRef.injector.get(ChangeDetectorRef).markForCheck();
        }
    }
    /**
     * @return {?}
     */
    ngDoCheck() {
        if (this.isDirective(this.content)) {
            this.content.check();
        }
    }
}
PolymorpheusOutletComponent.decorators = [
    { type: Component, args: [{
                selector: 'polymorpheus-outlet, [polymorpheus-outlet]',
                template: "<ng-container\r\n    *ngIf=\"isTemplate(content) || isComponent(content)\"\r\n    #mutableContext=\"context\"\r\n    context\r\n    [ngTemplateOutlet]=\"getTemplate(content, component)\"\r\n    [ngTemplateOutletContext]=\"context\"\r\n>\r\n    <ng-template #component>\r\n        <ng-container\r\n            *ngIf=\"isComponent(content)\"\r\n            [ngComponentOutlet]=\"content.component\"\r\n            [ngComponentOutletInjector]=\"content | injector : mutableContext.context : injector\"\r\n        ></ng-container>\r\n    </ng-template>\r\n</ng-container>\r\n<ng-container *ngIf=\"primitive\">\r\n    <ng-container *ngIf=\"!template; else child\">{{primitive}}</ng-container>\r\n    <ng-template #child>\r\n        <ng-container\r\n            [ngTemplateOutlet]=\"template\"\r\n            [ngTemplateOutletContext]=\"{$implicit: primitive}\"\r\n        ></ng-container>\r\n    </ng-template>\r\n</ng-container>\r\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [':host { display: block; }']
            }] }
];
/** @nocollapse */
PolymorpheusOutletComponent.ctorParameters = () => [
    { type: Injector, decorators: [{ type: Inject, args: [Injector,] }] }
];
PolymorpheusOutletComponent.propDecorators = {
    content: [{ type: Input }],
    context: [{ type: Input }],
    template: [{ type: ContentChild, args: [TemplateRef,] }],
    outlet: [{ type: ViewChild, args: [NgComponentOutlet,] }]
};
if (false) {
    /** @type {?} */
    PolymorpheusOutletComponent.prototype.content;
    /** @type {?} */
    PolymorpheusOutletComponent.prototype.context;
    /** @type {?} */
    PolymorpheusOutletComponent.prototype.template;
    /** @type {?} */
    PolymorpheusOutletComponent.prototype.outlet;
    /** @type {?} */
    PolymorpheusOutletComponent.prototype.injector;
}
