// @angular/core/esm5/src/render3/fields.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { getClosureSafeProperty } from '../util/property';
export var NG_COMP_DEF = getClosureSafeProperty({ ɵcmp: getClosureSafeProperty });
export var NG_DIR_DEF = getClosureSafeProperty({ ɵdir: getClosureSafeProperty });
export var NG_PIPE_DEF = getClosureSafeProperty({ ɵpipe: getClosureSafeProperty });
export var NG_MOD_DEF = getClosureSafeProperty({ ɵmod: getClosureSafeProperty });
export var NG_LOC_ID_DEF = getClosureSafeProperty({ ɵloc: getClosureSafeProperty });
export var NG_FACTORY_DEF = getClosureSafeProperty({ ɵfac: getClosureSafeProperty });
/**
 * If a directive is diPublic, bloomAdd sets a property on the type with this constant as
 * the key and the directive's unique ID as the value. This allows us to map directives to their
 * bloom filter bit for DI.
 */
// TODO(misko): This is wrong. The NG_ELEMENT_ID should never be minified.
export var NG_ELEMENT_ID = getClosureSafeProperty({ __NG_ELEMENT_ID__: getClosureSafeProperty });
