// @angular/compiler-cli/src/ngtsc/partial_evaluator/src/result.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/partial_evaluator/src/result", ["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * A collection of publicly exported declarations from a module. Each declaration is evaluated
     * lazily upon request.
     */
    var ResolvedModule = /** @class */ (function () {
        function ResolvedModule(exports, evaluate) {
            this.exports = exports;
            this.evaluate = evaluate;
        }
        ResolvedModule.prototype.getExport = function (name) {
            if (!this.exports.has(name)) {
                return undefined;
            }
            return this.evaluate(this.exports.get(name));
        };
        ResolvedModule.prototype.getExports = function () {
            var _this = this;
            var map = new Map();
            this.exports.forEach(function (decl, name) {
                map.set(name, _this.evaluate(decl));
            });
            return map;
        };
        return ResolvedModule;
    }());
    exports.ResolvedModule = ResolvedModule;
    /**
     * A value member of an enumeration.
     *
     * Contains a `Reference` to the enumeration itself, and the name of the referenced member.
     */
    var EnumValue = /** @class */ (function () {
        function EnumValue(enumRef, name, resolved) {
            this.enumRef = enumRef;
            this.name = name;
            this.resolved = resolved;
        }
        return EnumValue;
    }());
    exports.EnumValue = EnumValue;
    /**
     * An implementation of a known function that can be statically evaluated.
     * It could be a built-in function or method (such as `Array.prototype.slice`) or a TypeScript
     * helper (such as `__spread`).
     */
    var KnownFn = /** @class */ (function () {
        function KnownFn() {
        }
        return KnownFn;
    }());
    exports.KnownFn = KnownFn;
});
