// @angular/cdk/esm5/testing/harness-environment.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __awaiter, __generator, __read, __spread, __values } from "tslib";
import { ComponentHarness, HarnessPredicate, } from './component-harness';
/**
 * Base harness environment class that can be extended to allow `ComponentHarness`es to be used in
 * different test environments (e.g. testbed, protractor, etc.). This class implements the
 * functionality of both a `HarnessLoader` and `LocatorFactory`. This class is generic on the raw
 * element type, `E`, used by the particular test environment.
 */
var HarnessEnvironment = /** @class */ (function () {
    function HarnessEnvironment(rawRootElement) {
        this.rawRootElement = rawRootElement;
        this.rootElement = this.createTestElement(rawRootElement);
    }
    // Implemented as part of the `LocatorFactory` interface.
    HarnessEnvironment.prototype.documentRootLocatorFactory = function () {
        return this.createEnvironment(this.getDocumentRoot());
    };
    // Implemented as part of the `LocatorFactory` interface.
    HarnessEnvironment.prototype.locatorFor = function () {
        var _this = this;
        var queries = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            queries[_i] = arguments[_i];
        }
        return function () { return _assertResultFound(_this._getAllHarnessesAndTestElements(queries), _getDescriptionForLocatorForQueries(queries)); };
    };
    // Implemented as part of the `LocatorFactory` interface.
    HarnessEnvironment.prototype.locatorForOptional = function () {
        var _this = this;
        var queries = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            queries[_i] = arguments[_i];
        }
        return function () { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, this._getAllHarnessesAndTestElements(queries)];
                case 1: return [2 /*return*/, (_a.sent())[0] || null];
            }
        }); }); };
    };
    // Implemented as part of the `LocatorFactory` interface.
    HarnessEnvironment.prototype.locatorForAll = function () {
        var _this = this;
        var queries = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            queries[_i] = arguments[_i];
        }
        return function () { return _this._getAllHarnessesAndTestElements(queries); };
    };
    // Implemented as part of the `LocatorFactory` interface.
    HarnessEnvironment.prototype.harnessLoaderFor = function (selector) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this.createEnvironment;
                        return [4 /*yield*/, _assertResultFound(this.getAllRawElements(selector), [_getDescriptionForHarnessLoaderQuery(selector)])];
                    case 1: return [2 /*return*/, _a.apply(this, [_b.sent()])];
                }
            });
        });
    };
    // Implemented as part of the `LocatorFactory` interface.
    HarnessEnvironment.prototype.harnessLoaderForOptional = function (selector) {
        return __awaiter(this, void 0, void 0, function () {
            var elements;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getAllRawElements(selector)];
                    case 1:
                        elements = _a.sent();
                        return [2 /*return*/, elements[0] ? this.createEnvironment(elements[0]) : null];
                }
            });
        });
    };
    // Implemented as part of the `LocatorFactory` interface.
    HarnessEnvironment.prototype.harnessLoaderForAll = function (selector) {
        return __awaiter(this, void 0, void 0, function () {
            var elements;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getAllRawElements(selector)];
                    case 1:
                        elements = _a.sent();
                        return [2 /*return*/, elements.map(function (element) { return _this.createEnvironment(element); })];
                }
            });
        });
    };
    // Implemented as part of the `HarnessLoader` interface.
    HarnessEnvironment.prototype.getHarness = function (query) {
        return this.locatorFor(query)();
    };
    // Implemented as part of the `HarnessLoader` interface.
    HarnessEnvironment.prototype.getAllHarnesses = function (query) {
        return this.locatorForAll(query)();
    };
    // Implemented as part of the `HarnessLoader` interface.
    HarnessEnvironment.prototype.getChildLoader = function (selector) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this.createEnvironment;
                        return [4 /*yield*/, _assertResultFound(this.getAllRawElements(selector), [_getDescriptionForHarnessLoaderQuery(selector)])];
                    case 1: return [2 /*return*/, _a.apply(this, [_b.sent()])];
                }
            });
        });
    };
    // Implemented as part of the `HarnessLoader` interface.
    HarnessEnvironment.prototype.getAllChildLoaders = function (selector) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getAllRawElements(selector)];
                    case 1: return [2 /*return*/, (_a.sent()).map(function (e) { return _this.createEnvironment(e); })];
                }
            });
        });
    };
    /** Creates a `ComponentHarness` for the given harness type with the given raw host element. */
    HarnessEnvironment.prototype.createComponentHarness = function (harnessType, element) {
        return new harnessType(this.createEnvironment(element));
    };
    /**
     * Matches the given raw elements with the given list of element and harness queries to produce a
     * list of matched harnesses and test elements.
     */
    HarnessEnvironment.prototype._getAllHarnessesAndTestElements = function (queries) {
        return __awaiter(this, void 0, void 0, function () {
            var _a, allQueries, harnessQueries, elementQueries, harnessTypes, rawElements, skipSelectorCheck, perElementMatches;
            var _b;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        _a = _parseQueries(queries), allQueries = _a.allQueries, harnessQueries = _a.harnessQueries, elementQueries = _a.elementQueries, harnessTypes = _a.harnessTypes;
                        return [4 /*yield*/, this.getAllRawElements(__spread(elementQueries, harnessQueries.map(function (predicate) { return predicate.getSelector(); })).join(','))];
                    case 1:
                        rawElements = _c.sent();
                        skipSelectorCheck = (elementQueries.length === 0 && harnessTypes.size === 1) ||
                            harnessQueries.length === 0;
                        return [4 /*yield*/, Promise.all(rawElements.map(function (rawElement) { return __awaiter(_this, void 0, void 0, function () {
                                var testElement, allResultsForElement;
                                var _this = this;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            testElement = this.createTestElement(rawElement);
                                            return [4 /*yield*/, Promise.all(
                                                // For each query, get `null` if it doesn't match, or a `TestElement` or
                                                // `ComponentHarness` as appropriate if it does match. This gives us everything that
                                                // matches the current raw element, but it may contain duplicate entries (e.g. multiple
                                                // `TestElement` or multiple `ComponentHarness` of the same type.
                                                allQueries.map(function (query) {
                                                    return _this._getQueryResultForElement(query, rawElement, testElement, skipSelectorCheck);
                                                }))];
                                        case 1:
                                            allResultsForElement = _a.sent();
                                            return [2 /*return*/, _removeDuplicateQueryResults(allResultsForElement)];
                                    }
                                });
                            }); }))];
                    case 2:
                        perElementMatches = _c.sent();
                        return [2 /*return*/, (_b = []).concat.apply(_b, __spread(perElementMatches))];
                }
            });
        });
    };
    /**
     * Check whether the given query matches the given element, if it does return the matched
     * `TestElement` or `ComponentHarness`, if it does not, return null. In cases where the caller
     * knows for sure that the query matches the element's selector, `skipSelectorCheck` can be used
     * to skip verification and optimize performance.
     */
    HarnessEnvironment.prototype._getQueryResultForElement = function (query, rawElement, testElement, skipSelectorCheck) {
        if (skipSelectorCheck === void 0) { skipSelectorCheck = false; }
        return __awaiter(this, void 0, void 0, function () {
            var _a, _b, harness;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        if (!(typeof query === 'string')) return [3 /*break*/, 3];
                        _a = skipSelectorCheck;
                        if (_a) return [3 /*break*/, 2];
                        return [4 /*yield*/, testElement.matchesSelector(query)];
                    case 1:
                        _a = (_c.sent());
                        _c.label = 2;
                    case 2: return [2 /*return*/, ((_a) ? testElement : null)];
                    case 3:
                        _b = skipSelectorCheck;
                        if (_b) return [3 /*break*/, 5];
                        return [4 /*yield*/, testElement.matchesSelector(query.getSelector())];
                    case 4:
                        _b = (_c.sent());
                        _c.label = 5;
                    case 5:
                        if (!_b) return [3 /*break*/, 7];
                        harness = this.createComponentHarness(query.harnessType, rawElement);
                        return [4 /*yield*/, query.evaluate(harness)];
                    case 6: return [2 /*return*/, (_c.sent()) ? harness : null];
                    case 7: return [2 /*return*/, null];
                }
            });
        });
    };
    return HarnessEnvironment;
}());
export { HarnessEnvironment };
/**
 * Parses a list of queries in the format accepted by the `locatorFor*` methods into an easier to
 * work with format.
 */
function _parseQueries(queries) {
    var e_1, _a;
    var allQueries = [];
    var harnessQueries = [];
    var elementQueries = [];
    var harnessTypes = new Set();
    try {
        for (var queries_1 = __values(queries), queries_1_1 = queries_1.next(); !queries_1_1.done; queries_1_1 = queries_1.next()) {
            var query = queries_1_1.value;
            if (typeof query === 'string') {
                allQueries.push(query);
                elementQueries.push(query);
            }
            else {
                var predicate = query instanceof HarnessPredicate ? query : new HarnessPredicate(query, {});
                allQueries.push(predicate);
                harnessQueries.push(predicate);
                harnessTypes.add(predicate.harnessType);
            }
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (queries_1_1 && !queries_1_1.done && (_a = queries_1.return)) _a.call(queries_1);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return { allQueries: allQueries, harnessQueries: harnessQueries, elementQueries: elementQueries, harnessTypes: harnessTypes };
}
/**
 * Removes duplicate query results for a particular element. (e.g. multiple `TestElement`
 * instances or multiple instances of the same `ComponentHarness` class.
 */
function _removeDuplicateQueryResults(results) {
    return __awaiter(this, void 0, void 0, function () {
        var testElementMatched, matchedHarnessTypes, dedupedMatches, results_1, results_1_1, result;
        var e_2, _a;
        return __generator(this, function (_b) {
            testElementMatched = false;
            matchedHarnessTypes = new Set();
            dedupedMatches = [];
            try {
                for (results_1 = __values(results), results_1_1 = results_1.next(); !results_1_1.done; results_1_1 = results_1.next()) {
                    result = results_1_1.value;
                    if (!result) {
                        continue;
                    }
                    if (result instanceof ComponentHarness) {
                        if (!matchedHarnessTypes.has(result.constructor)) {
                            matchedHarnessTypes.add(result.constructor);
                            dedupedMatches.push(result);
                        }
                    }
                    else if (!testElementMatched) {
                        testElementMatched = true;
                        dedupedMatches.push(result);
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (results_1_1 && !results_1_1.done && (_a = results_1.return)) _a.call(results_1);
                }
                finally { if (e_2) throw e_2.error; }
            }
            return [2 /*return*/, dedupedMatches];
        });
    });
}
/** Verifies that there is at least one result in an array. */
function _assertResultFound(results, queryDescriptions) {
    return __awaiter(this, void 0, void 0, function () {
        var result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, results];
                case 1:
                    result = (_a.sent())[0];
                    if (result == undefined) {
                        throw Error("Failed to find element matching one of the following queries:\n" +
                            queryDescriptions.map(function (desc) { return "(" + desc + ")"; }).join(',\n'));
                    }
                    return [2 /*return*/, result];
            }
        });
    });
}
/** Gets a list of description strings from a list of queries. */
function _getDescriptionForLocatorForQueries(queries) {
    return queries.map(function (query) { return typeof query === 'string' ?
        _getDescriptionForTestElementQuery(query) : _getDescriptionForComponentHarnessQuery(query); });
}
/** Gets a description string for a `ComponentHarness` query. */
function _getDescriptionForComponentHarnessQuery(query) {
    var harnessPredicate = query instanceof HarnessPredicate ? query : new HarnessPredicate(query, {});
    var _a = harnessPredicate.harnessType, name = _a.name, hostSelector = _a.hostSelector;
    var description = name + " with host element matching selector: \"" + hostSelector + "\"";
    var constraints = harnessPredicate.getDescription();
    return description + (constraints ?
        " satisfying the constraints: " + harnessPredicate.getDescription() : '');
}
/** Gets a description string for a `TestElement` query. */
function _getDescriptionForTestElementQuery(selector) {
    return "TestElement for element matching selector: \"" + selector + "\"";
}
/** Gets a description string for a `HarnessLoader` query. */
function _getDescriptionForHarnessLoaderQuery(selector) {
    return "HarnessLoader for element matching selector: \"" + selector + "\"";
}
