// @angular/cdk/esm2015/a11y/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: src/cdk/a11y/public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
export { MESSAGES_CONTAINER_ID, CDK_DESCRIBEDBY_ID_PREFIX, CDK_DESCRIBEDBY_HOST_ATTRIBUTE, AriaDescriber } from './aria-describer/aria-describer';
export { ActiveDescendantKeyManager } from './key-manager/activedescendant-key-manager';
export { FocusKeyManager } from './key-manager/focus-key-manager';
export { ListKeyManager } from './key-manager/list-key-manager';
export { ConfigurableFocusTrap } from './focus-trap/configurable-focus-trap';
export { EventListenerFocusTrapInertStrategy } from './focus-trap/event-listener-inert-strategy';
export { FocusTrap, FocusTrapFactory, CdkTrapFocus } from './focus-trap/focus-trap';
export { ConfigurableFocusTrapFactory } from './focus-trap/configurable-focus-trap-factory';
export { FOCUS_TRAP_INERT_STRATEGY } from './focus-trap/focus-trap-inert-strategy';
export { InteractivityChecker } from './interactivity-checker/interactivity-checker';
export { LiveAnnouncer, CdkAriaLive } from './live-announcer/live-announcer';
export { LIVE_ANNOUNCER_ELEMENT_TOKEN_FACTORY, LIVE_ANNOUNCER_ELEMENT_TOKEN, LIVE_ANNOUNCER_DEFAULT_OPTIONS } from './live-announcer/live-announcer-tokens';
export { TOUCH_BUFFER_MS, FOCUS_MONITOR_DEFAULT_OPTIONS, FocusMonitor, CdkMonitorFocus } from './focus-monitor/focus-monitor';
export { isFakeMousedownFromScreenReader } from './fake-mousedown';
export { A11yModule } from './a11y-module';
export { HighContrastModeDetector, } from './high-contrast-mode/high-contrast-mode-detector';
