// @ng-web-apis/universal/esm5/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Public API Surface of @ng-web-apis/universal
 */
// Constants
export { UNIVERSAL_ANIMATION_FRAME } from './constants/universal-animation-frame';
export { UNIVERSAL_LOCAL_STORAGE } from './constants/universal-local-storage';
export { UNIVERSAL_LOCATION } from './constants/universal-location';
export { navigatorFactory, UNIVERSAL_NAVIGATOR } from './constants/universal-navigator';
export { performanceFactory, UNIVERSAL_PERFORMANCE } from './constants/universal-performance';
export { UNIVERSAL_SESSION_STORAGE } from './constants/universal-session-storage';
export { UNIVERSAL_SPEECH_SYNTHESIS } from './constants/universal-speech-synthesis';
export { UNIVERSAL_PROVIDERS, UNIVERSAL_TOKENS } from './constants/universal-tokens';
export { UNIVERSAL_USER_AGENT } from './constants/universal-user-agent';
// Utils
export { provideLocation } from './utils/provide-location';
export { provideUserAgent } from './utils/provide-user-agent';
