// @ng-web-apis/mutation-observer/esm5/directives/mutation-observer.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Attribute, Directive, ElementRef, EventEmitter, Inject, Output, } from '@angular/core';
import { MUTATION_OBSERVER_INIT } from '../tokens/mutation-observer-init';
import { mutationObserverInitFactory } from '../utils/mutation-observer-init-factory';
var ɵ0 = mutationObserverInitFactory;
// @dynamic
var MutationObserverDirective = /** @class */ (function (_super) {
    tslib_1.__extends(MutationObserverDirective, _super);
    function MutationObserverDirective(_a, config, _1, _2, _3, _4, _5, _6, _7) {
        var nativeElement = _a.nativeElement;
        var _this = _super.call(this, function (records) {
            _this.waMutationObserver.emit(records);
        }) || this;
        _this.waMutationObserver = new EventEmitter();
        _this.observe(nativeElement, config);
        return _this;
    }
    /**
     * @return {?}
     */
    MutationObserverDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.disconnect();
    };
    MutationObserverDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[waMutationObserver]',
                    exportAs: 'MutationObserver',
                    providers: [
                        {
                            provide: MUTATION_OBSERVER_INIT,
                            deps: [ElementRef],
                            useFactory: ɵ0,
                        },
                    ],
                },] }
    ];
    /** @nocollapse */
    MutationObserverDirective.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [MUTATION_OBSERVER_INIT,] }] },
        { type: undefined, decorators: [{ type: Attribute, args: ['attributeFilter',] }] },
        { type: undefined, decorators: [{ type: Attribute, args: ['attributeOldValue',] }] },
        { type: undefined, decorators: [{ type: Attribute, args: ['attributes',] }] },
        { type: undefined, decorators: [{ type: Attribute, args: ['characterData',] }] },
        { type: undefined, decorators: [{ type: Attribute, args: ['characterDataOldValue',] }] },
        { type: undefined, decorators: [{ type: Attribute, args: ['childList',] }] },
        { type: undefined, decorators: [{ type: Attribute, args: ['subtree',] }] }
    ]; };
    MutationObserverDirective.propDecorators = {
        waMutationObserver: [{ type: Output }]
    };
    return MutationObserverDirective;
}(MutationObserver));
export { MutationObserverDirective };
if (false) {
    /** @type {?} */
    MutationObserverDirective.prototype.waMutationObserver;
}
export { ɵ0 };
