// @angular/cdk/esm2015/scrolling/public-api.js
/**
 * @fileoverview added by tsickle
 * Generated from: src/cdk/scrolling/public-api.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
export { _fixedSizeVirtualScrollStrategyFactory, FixedSizeVirtualScrollStrategy, CdkFixedSizeVirtualScroll } from './fixed-size-virtual-scroll';
export { DEFAULT_SCROLL_TIME, ScrollDispatcher } from './scroll-dispatcher';
export { CdkScrollable } from './scrollable';
export { CdkScrollableModule, ScrollingModule } from './scrolling-module';
export { DEFAULT_RESIZE_TIME, ViewportRuler } from './viewport-ruler';
export { CdkVirtualForOf } from './virtual-for-of';
export { VIRTUAL_SCROLL_STRATEGY } from './virtual-scroll-strategy';
export { CdkVirtualScrollViewport } from './virtual-scroll-viewport';
