// @ng-web-apis/universal/esm5/constants/universal-navigator.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Optional } from '@angular/core';
import { NAVIGATOR } from '@ng-web-apis/common';
import { SSR_USER_AGENT } from '../tokens/ssr-user-agent';
import { alwaysFalse, alwaysRejected, alwaysZero, empty, emptyArray, emptyObject, } from '../utils/functions';
/** @type {?} */
var EVENT_TARGET = {
    addEventListener: empty,
    dispatchEvent: alwaysFalse,
    removeEventListener: empty,
};
/** @type {?} */
var PLUGIN = new (/** @class */ (function (_super) {
    tslib_1.__extends(class_1, _super);
    function class_1() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.description = '';
        _this.filename = '';
        _this.name = '';
        _this.version = '';
        return _this;
    }
    /**
     * @return {?}
     */
    class_1.prototype.item = /**
     * @return {?}
     */
    function () {
        return {
            description: '',
            enabledPlugin: this,
            suffixes: '',
            type: '',
        };
    };
    /**
     * @return {?}
     */
    class_1.prototype.namedItem = /**
     * @return {?}
     */
    function () {
        return {
            description: '',
            enabledPlugin: this,
            suffixes: '',
            type: '',
        };
    };
    return class_1;
}(Array)))();
/** @type {?} */
var MIME_TYPES = new (/** @class */ (function (_super) {
    tslib_1.__extends(class_2, _super);
    function class_2() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    class_2.prototype.item = /**
     * @return {?}
     */
    function () {
        return PLUGIN;
    };
    /**
     * @return {?}
     */
    class_2.prototype.namedItem = /**
     * @return {?}
     */
    function () {
        return PLUGIN;
    };
    /**
     * @return {?}
     */
    class_2.prototype.refresh = /**
     * @return {?}
     */
    function () { };
    return class_2;
}(Array)))();
/**
 * For older version of TS and Angular that do not support all properties from Navigator
 * @record
 */
function NavigatorLike() { }
/**
 * @param {?} userAgent
 * @return {?}
 */
export function navigatorFactory(userAgent) {
    return {
        appCodeName: '',
        appName: '',
        appVersion: '',
        platform: '',
        product: '',
        productSub: '',
        userAgent: userAgent || '',
        vendor: '',
        vendorSub: '',
        onLine: false,
        confirmSiteSpecificTrackingException: alwaysFalse,
        confirmWebWideTrackingException: alwaysFalse,
        removeSiteSpecificTrackingException: empty,
        removeWebWideTrackingException: empty,
        storeSiteSpecificTrackingException: empty,
        storeWebWideTrackingException: empty,
        msSaveBlob: alwaysFalse,
        msSaveOrOpenBlob: alwaysFalse,
        sendBeacon: alwaysFalse,
        hardwareConcurrency: 0,
        getDisplayMedia: alwaysRejected,
        language: '',
        languages: [],
        storage: {
            estimate: alwaysRejected,
            persist: alwaysRejected,
            persisted: alwaysRejected,
        },
        activeVRDisplays: [],
        authentication: {
            getAssertion: alwaysRejected,
            makeCredential: alwaysRejected,
        },
        clipboard: tslib_1.__assign({}, EVENT_TARGET, { readText: alwaysRejected, writeText: alwaysRejected }),
        cookieEnabled: false,
        doNotTrack: null,
        gamepadInputEmulation: 'keyboard',
        geolocation: {
            clearWatch: empty,
            getCurrentPosition: empty,
            watchPosition: alwaysZero,
        },
        maxTouchPoints: 0,
        mediaDevices: tslib_1.__assign({}, EVENT_TARGET, { ondevicechange: null, enumerateDevices: alwaysRejected, getSupportedConstraints: emptyObject, getUserMedia: alwaysRejected }),
        mimeTypes: MIME_TYPES,
        msManipulationViewsEnabled: false,
        msMaxTouchPoints: 0,
        msPointerEnabled: false,
        permissions: {
            query: alwaysRejected,
        },
        plugins: MIME_TYPES,
        pointerEnabled: false,
        serviceWorker: tslib_1.__assign({}, EVENT_TARGET, { controller: null, oncontrollerchange: null, onmessage: null, onmessageerror: null, ready: alwaysRejected(), getRegistration: alwaysRejected, getRegistrations: alwaysRejected, register: alwaysRejected, startMessages: empty }),
        webdriver: false,
        getGamepads: emptyArray,
        getUserMedia: empty,
        getVRDisplays: alwaysRejected,
        javaEnabled: alwaysFalse,
        msLaunchUri: empty,
        requestMediaKeySystemAccess: alwaysRejected,
        vibrate: alwaysFalse,
    };
}
/** @type {?} */
export var UNIVERSAL_NAVIGATOR = {
    provide: NAVIGATOR,
    deps: [[new Optional(), SSR_USER_AGENT]],
    useFactory: navigatorFactory,
};
