// @angular/compiler-cli/src/ngtsc/reflection/index.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/reflection", ["require", "exports", "tslib", "@angular/compiler-cli/src/ngtsc/reflection/src/host", "@angular/compiler-cli/src/ngtsc/reflection/src/type_to_value", "@angular/compiler-cli/src/ngtsc/reflection/src/typescript", "@angular/compiler-cli/src/ngtsc/reflection/src/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var tslib_1 = require("tslib");
    tslib_1.__exportStar(require("@angular/compiler-cli/src/ngtsc/reflection/src/host"), exports);
    var type_to_value_1 = require("@angular/compiler-cli/src/ngtsc/reflection/src/type_to_value");
    exports.typeNodeToValueExpr = type_to_value_1.typeNodeToValueExpr;
    var typescript_1 = require("@angular/compiler-cli/src/ngtsc/reflection/src/typescript");
    exports.TypeScriptReflectionHost = typescript_1.TypeScriptReflectionHost;
    exports.filterToMembersWithDecorator = typescript_1.filterToMembersWithDecorator;
    exports.reflectIdentifierOfDeclaration = typescript_1.reflectIdentifierOfDeclaration;
    exports.reflectNameOfDeclaration = typescript_1.reflectNameOfDeclaration;
    exports.reflectObjectLiteral = typescript_1.reflectObjectLiteral;
    exports.reflectTypeEntityToDeclaration = typescript_1.reflectTypeEntityToDeclaration;
    var util_1 = require("@angular/compiler-cli/src/ngtsc/reflection/src/util");
    exports.isNamedClassDeclaration = util_1.isNamedClassDeclaration;
    exports.isNamedFunctionDeclaration = util_1.isNamedFunctionDeclaration;
    exports.isNamedVariableDeclaration = util_1.isNamedVariableDeclaration;
});
