// @ng-web-apis/resize-observer/fesm2015/ng-web-apis-resize-observer.js
import { Observable } from 'rxjs';
import { share } from 'rxjs/operators';
import { WINDOW } from '@ng-web-apis/common';
import { InjectionToken, inject, ElementRef, Inject, Injectable, NgZone, Attribute, Directive, Output, NgModule } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const RESIZE_OPTION_BOX_DEFAULT = 'content-box';
/** @type {?} */
const RESIZE_OPTION_BOX = new InjectionToken('Box model to observe changes', {
    providedIn: 'root',
    factory: () => RESIZE_OPTION_BOX_DEFAULT,
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const RESIZE_OBSERVER_SUPPORT = new InjectionToken('Resize Observer API support', {
    providedIn: 'root',
    factory: () => !!inject(WINDOW).ResizeObserver,
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// @dynamic
class ResizeObserverService extends Observable {
    /**
     * @param {?} __0
     * @param {?} ngZone
     * @param {?} support
     * @param {?} box
     */
    constructor({ nativeElement }, ngZone, support, box) {
        /** @type {?} */
        let observer;
        super(subscriber => {
            if (!support) {
                subscriber.error('ResizeObserver is not supported in your browser');
                return;
            }
            observer = new ResizeObserver(entries => {
                ngZone.run(() => {
                    subscriber.next(entries);
                });
            });
            observer.observe(nativeElement, { box });
            return () => {
                observer.disconnect();
            };
        });
        return this.pipe(share());
    }
}
ResizeObserverService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ResizeObserverService.ctorParameters = () => [
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: NgZone, decorators: [{ type: Inject, args: [NgZone,] }] },
    { type: Boolean, decorators: [{ type: Inject, args: [RESIZE_OBSERVER_SUPPORT,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [RESIZE_OPTION_BOX,] }] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// TODO switch to Attribute once https://github.com/angular/angular/issues/36479 is fixed
/**
 * @param {?} __0
 * @return {?}
 */
function boxExtractor({ nativeElement, }) {
    /** @type {?} */
    const attribute = (/** @type {?} */ (nativeElement.getAttribute('waResizeBox')));
    return boxFactory(attribute);
}
/**
 * @param {?} box
 * @return {?}
 */
function boxFactory(box) {
    return box || RESIZE_OPTION_BOX_DEFAULT;
}
// @dynamic
class ResizeObserverDirective {
    /**
     * @param {?} entries$
     * @param {?} _box
     */
    constructor(entries$, _box) {
        this.waResizeObserver = entries$;
    }
}
ResizeObserverDirective.decorators = [
    { type: Directive, args: [{
                selector: '[waResizeObserver]',
                providers: [
                    ResizeObserverService,
                    {
                        provide: RESIZE_OPTION_BOX,
                        deps: [ElementRef],
                        useFactory: boxExtractor,
                    },
                ],
            },] }
];
/** @nocollapse */
ResizeObserverDirective.ctorParameters = () => [
    { type: Observable, decorators: [{ type: Inject, args: [ResizeObserverService,] }] },
    { type: undefined, decorators: [{ type: Attribute, args: ['waResizeBox',] }] }
];
ResizeObserverDirective.propDecorators = {
    waResizeObserver: [{ type: Output }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class ResizeObserverModule {
}
ResizeObserverModule.decorators = [
    { type: NgModule, args: [{
                declarations: [ResizeObserverDirective],
                exports: [ResizeObserverDirective],
            },] }
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { boxExtractor, boxFactory, ResizeObserverDirective, ResizeObserverModule, ResizeObserverService, RESIZE_OPTION_BOX_DEFAULT, RESIZE_OPTION_BOX, RESIZE_OBSERVER_SUPPORT };
