// @angular/cdk/schematics/ng-update/migrations/attribute-selectors.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/cdk/schematics/ng-update/migrations/attribute-selectors", ["require", "exports", "typescript", "@angular/cdk/schematics/update-tool/migration", "@angular/cdk/schematics/ng-update/typescript/literal", "@angular/cdk/schematics/ng-update/upgrade-data"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    const ts = require("typescript");
    const migration_1 = require("@angular/cdk/schematics/update-tool/migration");
    const literal_1 = require("@angular/cdk/schematics/ng-update/typescript/literal");
    const upgrade_data_1 = require("@angular/cdk/schematics/ng-update/upgrade-data");
    /**
     * Migration that walks through every string literal, template and stylesheet
     * in order to switch deprecated attribute selectors to the updated selector.
     */
    class AttributeSelectorsMigration extends migration_1.Migration {
        constructor() {
            super(...arguments);
            /** Required upgrade changes for specified target version. */
            this.data = upgrade_data_1.getVersionUpgradeData(this, 'attributeSelectors');
            // Only enable the migration rule if there is upgrade data.
            this.enabled = this.data.length !== 0;
        }
        visitNode(node) {
            if (ts.isStringLiteralLike(node)) {
                this._visitStringLiteralLike(node);
            }
        }
        visitTemplate(template) {
            this.data.forEach(selector => {
                literal_1.findAllSubstringIndices(template.content, selector.replace)
                    .map(offset => template.start + offset)
                    .forEach(start => this._replaceSelector(template.filePath, start, selector));
            });
        }
        visitStylesheet(stylesheet) {
            this.data.forEach(selector => {
                const currentSelector = `[${selector.replace}]`;
                const updatedSelector = `[${selector.replaceWith}]`;
                literal_1.findAllSubstringIndices(stylesheet.content, currentSelector)
                    .map(offset => stylesheet.start + offset)
                    .forEach(start => this._replaceSelector(stylesheet.filePath, start, { replace: currentSelector, replaceWith: updatedSelector }));
            });
        }
        _visitStringLiteralLike(literal) {
            if (literal.parent && literal.parent.kind !== ts.SyntaxKind.CallExpression) {
                return;
            }
            const literalText = literal.getText();
            const filePath = literal.getSourceFile().fileName;
            this.data.forEach(selector => {
                literal_1.findAllSubstringIndices(literalText, selector.replace)
                    .map(offset => literal.getStart() + offset)
                    .forEach(start => this._replaceSelector(filePath, start, selector));
            });
        }
        _replaceSelector(filePath, start, data) {
            this.fileSystem.edit(filePath)
                .remove(start, data.replace.length)
                .insertRight(start, data.replaceWith);
        }
    }
    exports.AttributeSelectorsMigration = AttributeSelectorsMigration;
});
