// rxjs-tslint-rules/dist/rules/rxjsNoCreateRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rule = void 0;
var tslib_1 = require("tslib");
var Lint = tslib_1.__importStar(require("tslint"));
var used_walker_1 = require("../support/used-walker");
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.applyWithProgram = function (sourceFile, program) {
        return this.applyWithWalker(new Walker(sourceFile, this.getOptions(), program));
    };
    Rule.metadata = {
        description: "Disallows the calling of Observable.create.",
        options: null,
        optionsDescription: "Not configurable.",
        requiresTypeInfo: true,
        ruleName: "rxjs-no-create",
        type: "functionality",
        typescriptOnly: true,
    };
    Rule.FAILURE_STRING = "Observable.create is forbidden; use new Observable.";
    return Rule;
}(Lint.Rules.TypedRule));
exports.Rule = Rule;
var Walker = (function (_super) {
    tslib_1.__extends(Walker, _super);
    function Walker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Walker.prototype.onSourceFileEnd = function () {
        var _this = this;
        if (this.usedStaticMethods["create"]) {
            this.usedStaticMethods["create"].forEach(function (node) {
                return _this.addFailureAtNode(node, Rule.FAILURE_STRING);
            });
        }
    };
    return Walker;
}(used_walker_1.UsedWalker));
