// rxjs-tslint-rules/dist/rules/rxjsPreferAngularAsyncPipeRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rule = void 0;
var tslib_1 = require("tslib");
var tsquery_1 = require("@phenomnomnominal/tsquery");
var Lint = tslib_1.__importStar(require("tslint"));
var peer = tslib_1.__importStar(require("../support/peer"));
var util_1 = require("../support/util");
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.applyWithProgram = function (sourceFile, program) {
        var _this = this;
        var failures = [];
        var typeChecker = program.getTypeChecker();
        var classDeclarations = tsquery_1.tsquery(sourceFile, "ClassDeclaration:has(Decorator[expression.expression.name=\"Component\"])");
        classDeclarations.forEach(function (classDeclaration) {
            var propertyAccessExpressions = tsquery_1.tsquery(classDeclaration, "CallExpression PropertyAccessExpression[name.name=\"subscribe\"]");
            propertyAccessExpressions.forEach(function (node) {
                var propertyAccessExpression = node;
                var type = typeChecker.getTypeAtLocation(propertyAccessExpression.expression);
                if (util_1.couldBeType(type, "Observable")) {
                    var name_1 = propertyAccessExpression.name;
                    failures.push(new Lint.RuleFailure(sourceFile, name_1.getStart(), name_1.getStart() + name_1.getWidth(), Rule.FAILURE_STRING, _this.ruleName));
                }
            });
        });
        return failures;
    };
    Rule.metadata = {
        deprecationMessage: peer.v5 ? peer.v5NotSupportedMessage : undefined,
        description: "Disallows the calling of `subscribe` within an Angular component.",
        options: null,
        optionsDescription: "Not configurable.",
        requiresTypeInfo: true,
        ruleName: "rxjs-prefer-angular-async-pipe",
        type: "style",
        typescriptOnly: true,
    };
    Rule.FAILURE_STRING = "Prefer async pipe over subscribe";
    return Rule;
}(Lint.Rules.TypedRule));
exports.Rule = Rule;
