// @angular/forms/esm2015/src/directives/form_interface.js
/**
 * @fileoverview added by tsickle
 * Generated from: packages/forms/src/directives/form_interface.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
 * \@description
 * An interface implemented by `FormGroupDirective` and `NgForm` directives.
 *
 * Only used by the `ReactiveFormsModule` and `FormsModule`.
 *
 * \@publicApi
 * @record
 */
export function Form() { }
if (false) {
    /**
     * \@description
     * Add a control to this form.
     *
     * @param {?} dir The control directive to add to the form.
     * @return {?}
     */
    Form.prototype.addControl = function (dir) { };
    /**
     * \@description
     * Remove a control from this form.
     *
     * @param {?} dir
     * @return {?}
     */
    Form.prototype.removeControl = function (dir) { };
    /**
     * \@description
     * The control directive from which to get the `FormControl`.
     *
     * @param {?} dir
     * @return {?}
     */
    Form.prototype.getControl = function (dir) { };
    /**
     * \@description
     * Add a group of controls to this form.
     *
     * @param {?} dir
     * @return {?}
     */
    Form.prototype.addFormGroup = function (dir) { };
    /**
     * \@description
     * Remove a group of controls to this form.
     *
     * @param {?} dir
     * @return {?}
     */
    Form.prototype.removeFormGroup = function (dir) { };
    /**
     * \@description
     * The `FormGroup` associated with a particular `AbstractFormGroupDirective`.
     *
     * @param {?} dir
     * @return {?}
     */
    Form.prototype.getFormGroup = function (dir) { };
    /**
     * \@description
     * Update the model for a particular control with a new value.
     *
     * @param {?} dir
     * @param {?} value
     * @return {?}
     */
    Form.prototype.updateModel = function (dir, value) { };
}
