// codelyzer/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var angularWhitespaceRule_1 = require("./angularWhitespaceRule");
exports.AngularWhitespaceRule = angularWhitespaceRule_1.Rule;
var componentClassSuffixRule_1 = require("./componentClassSuffixRule");
exports.ComponentClassSuffixRule = componentClassSuffixRule_1.Rule;
var componentMaxInlineDeclarationsRule_1 = require("./componentMaxInlineDeclarationsRule");
exports.ComponentMaxInlineDeclarationsRule = componentMaxInlineDeclarationsRule_1.Rule;
var componentSelectorRule_1 = require("./componentSelectorRule");
exports.ComponentSelectorRule = componentSelectorRule_1.Rule;
var contextualDecoratorRule_1 = require("./contextualDecoratorRule");
exports.ContextualDecoratorRule = contextualDecoratorRule_1.Rule;
var contextualLifecycleRule_1 = require("./contextualLifecycleRule");
exports.ContextualLifecycleRule = contextualLifecycleRule_1.Rule;
var directiveClassSuffixRule_1 = require("./directiveClassSuffixRule");
exports.DirectiveClassSuffixRule = directiveClassSuffixRule_1.Rule;
var directiveSelectorRule_1 = require("./directiveSelectorRule");
exports.DirectiveSelectorRule = directiveSelectorRule_1.Rule;
var importDestructuringSpacingRule_1 = require("./importDestructuringSpacingRule");
exports.ImportDestructuringSpacingRule = importDestructuringSpacingRule_1.Rule;
var noAttributeDecoratorRule_1 = require("./noAttributeDecoratorRule");
exports.NoAttributeDecoratorRule = noAttributeDecoratorRule_1.Rule;
var noConflictingLifecycleRule_1 = require("./noConflictingLifecycleRule");
exports.NoConflictingLifecycleRule = noConflictingLifecycleRule_1.Rule;
var noForwardRefRule_1 = require("./noForwardRefRule");
exports.NoForwardRefRule = noForwardRefRule_1.Rule;
var noHostMetadataPropertyRule_1 = require("./noHostMetadataPropertyRule");
exports.NoHostMetadataPropertyRule = noHostMetadataPropertyRule_1.Rule;
var noInputPrefixRule_1 = require("./noInputPrefixRule");
exports.NoInputPrefixRule = noInputPrefixRule_1.Rule;
var noInputRenameRule_1 = require("./noInputRenameRule");
exports.NoInputRenameRule = noInputRenameRule_1.Rule;
var noInputsMetadataPropertyRule_1 = require("./noInputsMetadataPropertyRule");
exports.NoInputsMetadataPropertyRule = noInputsMetadataPropertyRule_1.Rule;
var noLifecycleCallRule_1 = require("./noLifecycleCallRule");
exports.NoLifecycleCallRule = noLifecycleCallRule_1.Rule;
var noOutputNativeRule_1 = require("./noOutputNativeRule");
exports.NoOutputNativeRule = noOutputNativeRule_1.Rule;
var noOutputOnPrefixRule_1 = require("./noOutputOnPrefixRule");
exports.NoOutputOnPrefixRule = noOutputOnPrefixRule_1.Rule;
var noOutputRenameRule_1 = require("./noOutputRenameRule");
exports.NoOutputRenameRule = noOutputRenameRule_1.Rule;
var noOutputsMetadataPropertyRule_1 = require("./noOutputsMetadataPropertyRule");
exports.NoOutputsMetadataPropertyRule = noOutputsMetadataPropertyRule_1.Rule;
var noPipeImpureRule_1 = require("./noPipeImpureRule");
exports.NoPipeImpureRule = noPipeImpureRule_1.Rule;
var noQueriesMetadataPropertyRule_1 = require("./noQueriesMetadataPropertyRule");
exports.NoQueriesMetadataPropertyRule = noQueriesMetadataPropertyRule_1.Rule;
var noUnusedCssRule_1 = require("./noUnusedCssRule");
exports.NoUnusedCssRule = noUnusedCssRule_1.Rule;
var pipePrefixRule_1 = require("./pipePrefixRule");
exports.PipePrefixRule = pipePrefixRule_1.Rule;
var preferInlineDecoratorRule_1 = require("./preferInlineDecoratorRule");
exports.PreferInlineDecoratorRule = preferInlineDecoratorRule_1.Rule;
var preferOnPushComponentChangeDetectionRule_1 = require("./preferOnPushComponentChangeDetectionRule");
exports.PreferOnPushComponentChangeDetectionRule = preferOnPushComponentChangeDetectionRule_1.Rule;
var preferOutputReadonlyRule_1 = require("./preferOutputReadonlyRule");
exports.PreferOutputReadonlyRule = preferOutputReadonlyRule_1.Rule;
var relativeUrlPrefixRule_1 = require("./relativeUrlPrefixRule");
exports.RelativeUrlPrefixRule = relativeUrlPrefixRule_1.Rule;
var templateAccessibilityAltTextRule_1 = require("./templateAccessibilityAltTextRule");
exports.TemplateAccessibilityAltTextRule = templateAccessibilityAltTextRule_1.Rule;
var templateAccessibilityElementsContentRule_1 = require("./templateAccessibilityElementsContentRule");
exports.TemplateAccessibilityElementsContentRule = templateAccessibilityElementsContentRule_1.Rule;
var templateAccessibilityLabelForRule_1 = require("./templateAccessibilityLabelForRule");
exports.TemplateAccessibilityLabelForVisitor = templateAccessibilityLabelForRule_1.Rule;
var templateAccessibilityTabindexNoPositiveRule_1 = require("./templateAccessibilityTabindexNoPositiveRule");
exports.TemplateAccessibilityTabindexNoPositiveRule = templateAccessibilityTabindexNoPositiveRule_1.Rule;
var templateAccessibilityTableScopeRule_1 = require("./templateAccessibilityTableScopeRule");
exports.TemplateAccessibilityTableScopeRule = templateAccessibilityTableScopeRule_1.Rule;
var templateAccessibilityValidAriaRule_1 = require("./templateAccessibilityValidAriaRule");
exports.TemplateAccessibilityValidAriaRule = templateAccessibilityValidAriaRule_1.Rule;
var templateBananaInBoxRule_1 = require("./templateBananaInBoxRule");
exports.TemplateBananaInBoxRule = templateBananaInBoxRule_1.Rule;
var templateClickEventsHaveKeyEventsRule_1 = require("./templateClickEventsHaveKeyEventsRule");
exports.TemplateClickEventsHaveKeyEventsRule = templateClickEventsHaveKeyEventsRule_1.Rule;
var templateConditionalComplexityRule_1 = require("./templateConditionalComplexityRule");
exports.TemplateConditionalComplexityRule = templateConditionalComplexityRule_1.Rule;
var templateCyclomaticComplexityRule_1 = require("./templateCyclomaticComplexityRule");
exports.TemplateCyclomaticComplexityRule = templateCyclomaticComplexityRule_1.Rule;
var templateI18nRule_1 = require("./templateI18nRule");
exports.TemplateI18nRule = templateI18nRule_1.Rule;
var templateMouseEventsHaveKeyEventsRule_1 = require("./templateMouseEventsHaveKeyEventsRule");
exports.TemplateMouseEventsHaveKeyEventsRule = templateMouseEventsHaveKeyEventsRule_1.Rule;
var templateNoAnyRule_1 = require("./templateNoAnyRule");
exports.TemplateNoAnyRule = templateNoAnyRule_1.Rule;
var templateNoAutofocusRule_1 = require("./templateNoAutofocusRule");
exports.TemplateNoAutofocusRule = templateNoAutofocusRule_1.Rule;
var templateNoCallExpressionRule_1 = require("./templateNoCallExpressionRule");
exports.TemplateNoCallExpressionRule = templateNoCallExpressionRule_1.Rule;
var templateNoDistractingElementsRule_1 = require("./templateNoDistractingElementsRule");
exports.TemplateNoDistractingElementsRule = templateNoDistractingElementsRule_1.Rule;
var templateNoNegatedAsyncRule_1 = require("./templateNoNegatedAsyncRule");
exports.TemplateNoNegatedAsyncRule = templateNoNegatedAsyncRule_1.Rule;
var templateUseTrackByFunctionRule_1 = require("./templateUseTrackByFunctionRule");
exports.TemplateUseTrackByFunctionRule = templateUseTrackByFunctionRule_1.Rule;
var useComponentSelectorRule_1 = require("./useComponentSelectorRule");
exports.UseComponentSelectorRule = useComponentSelectorRule_1.Rule;
var useComponentViewEncapsulationRule_1 = require("./useComponentViewEncapsulationRule");
exports.UseComponentViewEncapsulationRule = useComponentViewEncapsulationRule_1.Rule;
var useInjectableProvidedInRule_1 = require("./useInjectableProvidedInRule");
exports.UseInjectableProvidedInRule = useInjectableProvidedInRule_1.Rule;
var useLifecycleInterfaceRule_1 = require("./useLifecycleInterfaceRule");
exports.UseLifecycleInterfaceRule = useLifecycleInterfaceRule_1.Rule;
var usePipeDecoratorRule_1 = require("./usePipeDecoratorRule");
exports.UsePipeDecoratorRule = usePipeDecoratorRule_1.Rule;
var usePipeTransformInterfaceRule_1 = require("./usePipeTransformInterfaceRule");
exports.UsePipeTransformInterfaceRule = usePipeTransformInterfaceRule_1.Rule;
__export(require("./angular"));
exports.rulesDirectory = '.';
