// @ng-web-apis/universal/esm5/classes/storage-mock.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
var StorageMock = /** @class */ (function () {
    function StorageMock() {
        this.storage = new Map();
    }
    Object.defineProperty(StorageMock.prototype, "length", {
        get: /**
         * @return {?}
         */
        function () {
            return this.storage.size;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} key
     * @return {?}
     */
    StorageMock.prototype.getItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return this.storage.has(key) ? (/** @type {?} */ (this.storage.get(key))) : null;
    };
    /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    StorageMock.prototype.setItem = /**
     * @param {?} key
     * @param {?} value
     * @return {?}
     */
    function (key, value) {
        this.storage.set(key, value);
    };
    /**
     * @return {?}
     */
    StorageMock.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.storage.clear();
    };
    /**
     * @param {?} index
     * @return {?}
     */
    StorageMock.prototype.key = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        return index < this.storage.size ? tslib_1.__spread(this.storage.keys())[index] : null;
    };
    /**
     * @param {?} key
     * @return {?}
     */
    StorageMock.prototype.removeItem = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        this.storage.delete(key);
    };
    return StorageMock;
}());
export { StorageMock };
if (false) {
    /**
     * @type {?}
     * @private
     */
    StorageMock.prototype.storage;
}
