// @ng-web-apis/common/fesm5/ng-web-apis-common.js
import { DOCUMENT } from '@angular/common';
import { animationFrameScheduler, of, fromEvent } from 'rxjs';
import { map, repeat, share, distinctUntilChanged, startWith } from 'rxjs/operators';
import { inject, InjectionToken } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var WINDOW = new InjectionToken('An abstraction over global window object', {
    factory: function () {
        var defaultView = inject(DOCUMENT).defaultView;
        if (!defaultView) {
            throw new Error('Window is not available');
        }
        return defaultView;
    },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var PERFORMANCE = new InjectionToken('An abstraction over window.performance object', {
    factory: function () { return inject(WINDOW).performance; },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
// interval has bug: https://github.com/ReactiveX/rxjs/issues/4972, scheduled only present in rxjs 6.5+
/** @type {?} */
var ANIMATION_FRAME = new InjectionToken('Shared Observable based on `window.requestAnimationFrame`', {
    factory: function () {
        /** @type {?} */
        var performanceRef = inject(PERFORMANCE);
        return of(0, animationFrameScheduler).pipe(repeat(), map(function () { return performanceRef.now(); }), share());
    },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var CSS = new InjectionToken('An abstraction over window.CSS object', {
    factory: function () {
        return ((/** @type {?} */ (inject(WINDOW)))).CSS || {
            escape: function (v) { return v; },
            supports: function () { return false; },
        };
    },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var LOCAL_STORAGE = new InjectionToken('An abstraction over window.localStorage object', {
    factory: function () { return inject(WINDOW).localStorage; },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var LOCATION = new InjectionToken('An abstraction over window.location object', {
    factory: function () { return inject(WINDOW).location; },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var NAVIGATOR = new InjectionToken('An abstraction over window.navigator object', {
    factory: function () { return inject(WINDOW).navigator; },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var NETWORK_INFORMATION = new InjectionToken('An abstraction over window.navigator.connection object', {
    factory: function () { return inject(NAVIGATOR).connection || null; },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var PAGE_VISIBILITY = new InjectionToken('Shared Observable based on `document visibility changed`', {
    factory: function () {
        /** @type {?} */
        var documentRef = inject(DOCUMENT);
        return fromEvent(documentRef, 'visibilitychange').pipe(startWith(0), map(function () { return !documentRef.hidden; }), distinctUntilChanged(), share());
    },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var SESSION_STORAGE = new InjectionToken('An abstraction over window.sessionStorage object', {
    factory: function () { return inject(WINDOW).sessionStorage; },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
var USER_AGENT = new InjectionToken('An abstraction over window.navigator.userAgent object', {
    factory: function () { return inject(NAVIGATOR).userAgent; },
});

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */

export { ANIMATION_FRAME, CSS, LOCAL_STORAGE, LOCATION, NAVIGATOR, NETWORK_INFORMATION, PAGE_VISIBILITY, PERFORMANCE, SESSION_STORAGE, USER_AGENT, WINDOW };
