// tslint-eslint-rules/dist/rules/arrayBracketSpacingRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var OPTION_ALWAYS = 'always';
var RULE_NAME = 'array-bracket-spacing';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        var walker = new ArrayBracketSpacingWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    };
    Rule.metadata = {
        ruleName: RULE_NAME,
        description: 'enforce consistent spacing inside array brackets',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      A number of style guides require or disallow spaces between array brackets and other tokens.\n      This rule applies to both array literals and destructuring assignments (ECMAScript 6).\n      "], ["\n      A number of style guides require or disallow spaces between array brackets and other tokens.\n      This rule applies to both array literals and destructuring assignments (ECMAScript 6).\n      "]))),
        optionsDescription: Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      The rule takes one or two options. The first is a string, which can be:\n\n      - `\"never\"` (default) disallows spaces inside array brackets\n      - `\"always\"`requires one or more spaces or newlines inside array brackets\n\n      The second option is an object for exceptions to the `\"never\"` option:\n\n      - `\"singleValue\": true` requires one or more spaces or newlines inside brackets of array\n                                literals that contain a single element\n      - `\"objectsInArrays\": true` requires one or more spaces or newlines between brackets of\n                                    array literals and braces of their object literal elements\n                                    `[ {` or `} ]`\n      - `\"arraysInArrays\": true` requires one or more spaces or newlines between brackets of\n                                   array literals and brackets of their array literal elements\n                                   `[ [` or `] ]`\n\n      When using the `\"always\"` option the second option takes on these exceptions:\n\n      - `\"singleValue\": false` disallows spaces inside brackets of array literals that contain a\n                                 single element\n      - `\"objectsInArrays\": false` disallows spaces between brackets of array literals and braces\n                                     of their object literal elements `[ {` or `} ]`\n      - `\"arraysInArrays\": false` disallows spaces between brackets of array literals and brackets\n                                    of their array literal elements `[ [` or `] ]`\n\n      This rule has build-in exceptions:\n\n      - `\"never\"` (and also the exceptions to the `\"always\"` option) allows newlines inside\n                    array brackets, because this is a common pattern\n      - `\"always\"` does not require spaces or newlines in empty array literals `[]`\n      "], ["\n      The rule takes one or two options. The first is a string, which can be:\n\n      - \\`\"never\"\\` (default) disallows spaces inside array brackets\n      - \\`\"always\"\\`requires one or more spaces or newlines inside array brackets\n\n      The second option is an object for exceptions to the \\`\"never\"\\` option:\n\n      - \\`\"singleValue\": true\\` requires one or more spaces or newlines inside brackets of array\n                                literals that contain a single element\n      - \\`\"objectsInArrays\": true\\` requires one or more spaces or newlines between brackets of\n                                    array literals and braces of their object literal elements\n                                    \\`[ {\\` or \\`} ]\\`\n      - \\`\"arraysInArrays\": true\\` requires one or more spaces or newlines between brackets of\n                                   array literals and brackets of their array literal elements\n                                   \\`[ [\\` or \\`] ]\\`\n\n      When using the \\`\"always\"\\` option the second option takes on these exceptions:\n\n      - \\`\"singleValue\": false\\` disallows spaces inside brackets of array literals that contain a\n                                 single element\n      - \\`\"objectsInArrays\": false\\` disallows spaces between brackets of array literals and braces\n                                     of their object literal elements \\`[ {\\` or \\`} ]\\`\n      - \\`\"arraysInArrays\": false\\` disallows spaces between brackets of array literals and brackets\n                                    of their array literal elements \\`[ [\\` or \\`] ]\\`\n\n      This rule has build-in exceptions:\n\n      - \\`\"never\"\\` (and also the exceptions to the \\`\"always\"\\` option) allows newlines inside\n                    array brackets, because this is a common pattern\n      - \\`\"always\"\\` does not require spaces or newlines in empty array literals \\`[]\\`\n      "]))),
        options: {
            anyOf: [
                {
                    type: 'array',
                    items: [
                        {
                            enum: ['always', 'never']
                        }
                    ],
                    minItems: 0,
                    maxItems: 1
                },
                {
                    type: 'object',
                    properties: {
                        singleValue: {
                            type: 'boolean'
                        },
                        objectsInArrays: {
                            type: 'boolean'
                        },
                        arraysInArrays: {
                            type: 'boolean'
                        }
                    },
                    additionalProperties: false
                }
            ]
        },
        optionExamples: [
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\"]\n        "], ["\n        \"", "\": [true, \"always\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"never\"]\n        "], ["\n        \"", "\": [true, \"never\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"never\", {\n          \"arraysInArrays\": true\n        }]\n        "], ["\n        \"", "\": [true, \"never\", {\n          \"arraysInArrays\": true\n        }]\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'style'
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var ArrayBracketSpacingWalker = (function (_super) {
    tslib_1.__extends(ArrayBracketSpacingWalker, _super);
    function ArrayBracketSpacingWalker(sourceFile, options) {
        var _this = _super.call(this, sourceFile, options) || this;
        _this.singleValueException = false;
        _this.objectsInArraysException = false;
        _this.arraysInArraysException = false;
        var ruleOptions = _this.getOptions();
        _this.spaced = _this.hasOption(OPTION_ALWAYS) || (ruleOptions && ruleOptions.length === 0);
        var opt = ruleOptions[1];
        var isDef = function (x) { return typeof x !== 'undefined'; };
        if (opt) {
            _this.singleValueException = isDef(opt.singleValue) && opt.singleValue !== _this.spaced;
            _this.objectsInArraysException = isDef(opt.objectsInArrays) && opt.objectsInArrays !== _this.spaced;
            _this.arraysInArraysException = isDef(opt.arraysInArrays) && opt.arraysInArrays !== _this.spaced;
        }
        return _this;
    }
    ArrayBracketSpacingWalker.prototype.report = function (start, msg, fix) {
        this.addFailure(this.createFailure(start, 1, msg, fix));
    };
    ArrayBracketSpacingWalker.prototype.reportNoBeginningSpace = function (token, space) {
        var start = token.getStart(this.getSourceFile());
        var fix = Lint.Replacement.deleteText(start + 1, space);
        this.report(start, 'There should be no space after "["', fix);
    };
    ArrayBracketSpacingWalker.prototype.reportRequiredBeginningSpace = function (token) {
        var start = token.getStart(this.getSourceFile());
        var fix = Lint.Replacement.appendText(start + 1, ' ');
        this.report(start, 'A space is required after "["', fix);
    };
    ArrayBracketSpacingWalker.prototype.reportRequiredEndingSpace = function (token) {
        var start = token.getStart(this.getSourceFile());
        var fix = Lint.Replacement.appendText(start, ' ');
        this.report(start, 'A space is required before "]"', fix);
    };
    ArrayBracketSpacingWalker.prototype.reportNoEndingSpace = function (token, space) {
        var start = token.getStart(this.getSourceFile());
        var fix = Lint.Replacement.deleteText(start - space, space);
        this.report(start, 'There should be no space before "]"', fix);
    };
    ArrayBracketSpacingWalker.prototype.visitNode = function (node) {
        if (node.kind === ts.SyntaxKind.ArrayBindingPattern) {
            this.validateArraySpacing(node, node.elements);
        }
        _super.prototype.visitNode.call(this, node);
    };
    ArrayBracketSpacingWalker.prototype.visitArrayLiteralExpression = function (node) {
        this.validateArraySpacing(node, node.elements);
        _super.prototype.visitArrayLiteralExpression.call(this, node);
    };
    ArrayBracketSpacingWalker.prototype.isObjectType = function (node) {
        return node && node.kind === ts.SyntaxKind.ObjectLiteralExpression;
    };
    ArrayBracketSpacingWalker.prototype.isArrayType = function (node) {
        if (node) {
            if (node.kind === ts.SyntaxKind.ArrayLiteralExpression) {
                return true;
            }
            var firstChild = node.getChildAt(0);
            if (firstChild && firstChild.kind === ts.SyntaxKind.ArrayBindingPattern) {
                return true;
            }
        }
        return false;
    };
    ArrayBracketSpacingWalker.prototype.validateArraySpacing = function (node, elements) {
        var _this = this;
        if (this.spaced && elements.length === 0) {
            return;
        }
        var first = node.getChildAt(0);
        var last = node.getChildAt(2);
        var firstElement = elements[0];
        var lastElement = elements[elements.length - 1];
        var second = firstElement || last;
        var penultimate = lastElement || first;
        if (second.pos === second.end) {
            second = node.getChildAt(1).getChildAt(1);
        }
        if (elements.hasTrailingComma) {
            penultimate = elements;
        }
        var mustBeSpaced = function (token) { return (_this.singleValueException && elements.length === 1 ||
            _this.objectsInArraysException && _this.isObjectType(token) ||
            _this.arraysInArraysException && _this.isArrayType(token)) ? !_this.spaced : _this.spaced; };
        var openingBracketMustBeSpaced = mustBeSpaced(firstElement);
        var closingBracketMustBeSpaced = mustBeSpaced(lastElement);
        var spaceAfterOpeningBracket = this.getSpaceBetween(first, second, false);
        var isBreakAfterOpeningBracket = this.isLineBreakBetween(first, second);
        var spaceBeforeClosingBracket = this.getSpaceBetween(penultimate, last, true);
        var isBreakBeforeClosingBracket = this.isLineBreakBetween(penultimate, last);
        if (!isBreakAfterOpeningBracket) {
            if (openingBracketMustBeSpaced && !spaceAfterOpeningBracket) {
                this.reportRequiredBeginningSpace(first);
            }
            else if (!openingBracketMustBeSpaced && spaceAfterOpeningBracket) {
                this.reportNoBeginningSpace(first, spaceAfterOpeningBracket);
            }
        }
        if (first !== penultimate && !isBreakBeforeClosingBracket) {
            if (closingBracketMustBeSpaced && !spaceBeforeClosingBracket) {
                this.reportRequiredEndingSpace(last);
            }
            else if (!closingBracketMustBeSpaced && spaceBeforeClosingBracket) {
                this.reportNoEndingSpace(last, spaceBeforeClosingBracket);
            }
        }
    };
    ArrayBracketSpacingWalker.prototype.getSpaceBetween = function (node, nextNode, trailing) {
        var end = nextNode.getStart(this.getSourceFile());
        var start = node.end;
        var text = this.getSourceFile().text.substring(start, end);
        var m = text.match(/\/\*.*\*\//);
        if (m && typeof m.index === 'number') {
            var len = m[0].length;
            return trailing ? end - (start + m.index + len) : m.index;
        }
        return end - start;
    };
    ArrayBracketSpacingWalker.prototype.isLineBreakBetween = function (node, nextNode) {
        return this.getEndPosition(node).line !== this.getStartPosition(nextNode).line;
    };
    ArrayBracketSpacingWalker.prototype.getStartPosition = function (node) {
        var srcFile = this.getSourceFile();
        return srcFile.getLineAndCharacterOfPosition(node.getStart(srcFile));
    };
    ArrayBracketSpacingWalker.prototype.getEndPosition = function (node) {
        return this.getSourceFile().getLineAndCharacterOfPosition(node.end);
    };
    return ArrayBracketSpacingWalker;
}(Lint.RuleWalker));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5;
