// tslint-eslint-rules/dist/rules/terArrowParensRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var RULE_NAME = 'ter-arrow-parens';
var always = 'Expected parentheses around arrow function argument.';
var asNeeded = 'Unexpected parentheses around single function argument.';
var block = 'Unexpected parentheses around single function argument having a body with no curly braces.';
var blockNoParens = 'Expected parentheses around arrow function argument having a body with curly braces.';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        var walker = new RuleWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    };
    Rule.metadata = {
        ruleName: RULE_NAME,
        description: 'require parens in arrow function arguments',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      Arrow functions can omit parentheses when they have exactly one parameter. In all other cases\n      the parameter(s) must be wrapped in parentheses. This rule enforces the consistent use of\n      parentheses in arrow functions.\n      "], ["\n      Arrow functions can omit parentheses when they have exactly one parameter. In all other cases\n      the parameter(s) must be wrapped in parentheses. This rule enforces the consistent use of\n      parentheses in arrow functions.\n      "]))),
        optionsDescription: Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      This rule has a string option and an object one.\n\n      String options are:\n\n      - `\"always\"` (default) requires parentheses around arguments in all cases.\n      - `\"as-needed\"` allows omitting parentheses when there is only one argument.\n\n      Object properties for variants of the `\"as-needed\"` option:\n\n      - `\"requireForBlockBody\": true` modifies the as-needed rule in order to require\n        parentheses if the function body is in an instructions block (surrounded by braces).\n      "], ["\n      This rule has a string option and an object one.\n\n      String options are:\n\n      - \\`\"always\"\\` (default) requires parentheses around arguments in all cases.\n      - \\`\"as-needed\"\\` allows omitting parentheses when there is only one argument.\n\n      Object properties for variants of the \\`\"as-needed\"\\` option:\n\n      - \\`\"requireForBlockBody\": true\\` modifies the as-needed rule in order to require\n        parentheses if the function body is in an instructions block (surrounded by braces).\n      "]))),
        options: {
            type: 'array',
            items: [
                {
                    enum: ['always', 'as-needed']
                },
                {
                    type: 'object',
                    properties: {
                        requireForBlockBody: {
                            type: 'boolean'
                        }
                    },
                    additionalProperties: false
                }
            ],
            maxLength: 1
        },
        optionExamples: [
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true]\n        "], ["\n        \"", "\": [true]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\"]\n        "], ["\n        \"", "\": [true, \"always\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"as-needed\"]\n        "], ["\n        \"", "\": [true, \"as-needed\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_6 || (templateObject_6 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"as-needed\", { \"requireForBlockBody\": true }]\n        "], ["\n        \"", "\": [true, \"as-needed\", { \"requireForBlockBody\": true }]\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'style'
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var RuleWalker = (function (_super) {
    tslib_1.__extends(RuleWalker, _super);
    function RuleWalker(sourceFile, options) {
        var _this = _super.call(this, sourceFile, options) || this;
        _this.srcFile = sourceFile;
        var opt = _this.getOptions();
        _this.asNeeded = opt[0] === 'as-needed';
        _this.requireForBlockBody = _this.asNeeded && opt[1] && opt[1].requireForBlockBody === true;
        return _this;
    }
    RuleWalker.prototype.visitArrowFunction = function (node) {
        _super.prototype.visitArrowFunction.call(this, node);
        if (node.parameters.length === 1) {
            var skip = Lint.hasModifier(node.modifiers, ts.SyntaxKind.AsyncKeyword) ? 1 : 0;
            var parameter = node.parameters[0];
            var text = parameter.getText();
            var firstToken = node.getChildAt(skip);
            var lastToken = node.getChildAt(2 + skip);
            var position = parameter.getStart();
            var paramWidth = text.length;
            var parensWidth = lastToken.end - firstToken.getStart(this.srcFile);
            var isGenerics = firstToken.kind === ts.SyntaxKind.LessThanToken;
            var hasParens = firstToken.kind === ts.SyntaxKind.OpenParenToken;
            var bodyIsBlock = node.body.kind === ts.SyntaxKind.Block;
            var isIdentifier = parameter.name.kind === ts.SyntaxKind.Identifier;
            var hasAnnotations = parameter.initializer || parameter.dotDotDotToken || parameter.type;
            var isSingleIdentifier = isIdentifier && !hasAnnotations;
            if (this.requireForBlockBody) {
                if (isSingleIdentifier && !node.type && !bodyIsBlock) {
                    if (hasParens) {
                        this.report(position - 1, parensWidth, block);
                    }
                    return;
                }
                if (bodyIsBlock && !isGenerics) {
                    if (!hasParens) {
                        this.report(position, paramWidth, blockNoParens);
                    }
                    return;
                }
            }
            if (this.asNeeded && isSingleIdentifier && !node.type) {
                if (hasParens) {
                    this.report(position - 1, parensWidth, asNeeded);
                }
                return;
            }
            if (!hasParens && !isGenerics) {
                this.report(position, paramWidth, always);
            }
        }
    };
    RuleWalker.prototype.report = function (position, width, message) {
        var failure = this.createFailure(position, width, message);
        this.addFailure(failure);
    };
    return RuleWalker;
}(Lint.RuleWalker));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
