// tslint-eslint-rules/dist/rules/terComputedPropertySpacingRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var RULE_NAME = 'ter-computed-property-spacing';
var ALWAYS_BEFORE_MESSAGE = "A space is required before ']'.";
var ALWAYS_AFTER_MESSAGE = "A space is required after '['.";
var NEVER_BEFORE_MESSAGE = "There should be no space before ']'.";
var NEVER_AFTER_MESSAGE = "There should be no space after '['.";
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.formatOptions = function (_a) {
        var alwaysOrNever = _a[0];
        return {
            always: alwaysOrNever === 'always'
        };
    };
    Rule.prototype.apply = function (sourceFile) {
        var opt = this.formatOptions(this.ruleArguments);
        var walker = new RuleWalker(sourceFile, this.ruleName, opt);
        return this.applyWithWalker(walker);
    };
    Rule.metadata = {
        ruleName: RULE_NAME,
        hasFix: true,
        description: 'require or disallow padding inside computed properties',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      While formatting preferences are very personal, a number of style guides require or disallow spaces between computed properties in the following situations:\n    "], ["\n      While formatting preferences are very personal, a number of style guides require or disallow spaces between computed properties in the following situations:\n    "]))),
        optionsDescription: Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      The rule takes in one option, which defines to require or forbid whitespace.\n\n      * `\"never\"` (default) disallows spaces inside computed property brackets\n      * `\"always\"` requires one or more spaces inside computed property brackets\n    "], ["\n      The rule takes in one option, which defines to require or forbid whitespace.\n\n      * \\`\"never\"\\` (default) disallows spaces inside computed property brackets\n      * \\`\"always\"\\` requires one or more spaces inside computed property brackets\n    "]))),
        options: {
            type: 'array',
            items: [{
                    enum: ['always', 'never']
                }],
            maxLength: 1
        },
        optionExamples: [
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true]\n        "], ["\n        \"", "\": [true]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\"]\n      "], ["\n        \"", "\": [true, \"always\"]\n      "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"never\"]\n      "], ["\n        \"", "\": [true, \"never\"]\n      "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'style'
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var RuleWalker = (function (_super) {
    tslib_1.__extends(RuleWalker, _super);
    function RuleWalker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RuleWalker.prototype.walk = function (sourceFile) {
        var _this = this;
        var cb = function (node) {
            if (node.kind === ts.SyntaxKind.ElementAccessExpression) {
                _this.checkNode(node, node.getChildAt(1), node.getChildAt(3));
            }
            else if (node.kind === ts.SyntaxKind.ComputedPropertyName) {
                _this.checkNode(node, node.getChildAt(0), node.getChildAt(2));
            }
            ts.forEachChild(node, cb);
        };
        ts.forEachChild(sourceFile, cb);
    };
    RuleWalker.prototype.checkNode = function (node, leftBracketNode, rightBracketNode) {
        var nodeText = node.getText();
        var regex = /\[([\s\S]*)\]/;
        var match = regex.exec(nodeText);
        if (!match) {
            return;
        }
        var contentWithinBrackets = match[1];
        if (this.options.always) {
            var beforeWhitespaceLength = this.getBeforeWhitespaceLength(contentWithinBrackets, true);
            var afterWhitespaceLength = this.getAfterWhitespaceLength(contentWithinBrackets, true);
            if (beforeWhitespaceLength === 0) {
                this.addFailureAtNode(leftBracketNode, ALWAYS_AFTER_MESSAGE, Lint.Replacement.appendText(leftBracketNode.getEnd(), ' '));
            }
            if (afterWhitespaceLength === 0) {
                this.addFailureAtNode(rightBracketNode, ALWAYS_BEFORE_MESSAGE, Lint.Replacement.appendText(rightBracketNode.getStart(), ' '));
            }
        }
        else {
            var contentWithinBracketsNoNewlines = contentWithinBrackets.replace('\n', '');
            var beforeWhitespaceLength = this.getBeforeWhitespaceLength(contentWithinBracketsNoNewlines, false);
            var afterWhitespaceLength = this.getAfterWhitespaceLength(contentWithinBracketsNoNewlines, false);
            if (beforeWhitespaceLength !== 0) {
                this.addFailureAtNode(leftBracketNode, NEVER_AFTER_MESSAGE, Lint.Replacement.deleteText(leftBracketNode.getEnd(), beforeWhitespaceLength));
            }
            if (afterWhitespaceLength !== 0) {
                this.addFailureAtNode(rightBracketNode, NEVER_BEFORE_MESSAGE, Lint.Replacement.deleteText(rightBracketNode.getStart() - afterWhitespaceLength, afterWhitespaceLength));
            }
        }
    };
    RuleWalker.prototype.getBeforeWhitespaceLength = function (content, newlinesCountAsWhitespace) {
        var regex = newlinesCountAsWhitespace ? /^\s+/ : /^[^\S\n]+/;
        var match = regex.exec(content);
        if (match) {
            return match[0].length;
        }
        else {
            return 0;
        }
    };
    RuleWalker.prototype.getAfterWhitespaceLength = function (content, newlinesCountAsWhitespace) {
        var regex = newlinesCountAsWhitespace ? /\s+$/ : /[^\S\n]+$/;
        var match = regex.exec(content);
        if (match) {
            return match[0].length;
        }
        else {
            return 0;
        }
    };
    return RuleWalker;
}(Lint.AbstractWalker));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5;
