// tslint-eslint-rules/dist/rules/terFuncCallSpacingRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var RULE_NAME = 'ter-func-call-spacing';
var ALWAYS = 'always';
var MISSING_SPACE = 'Missing space between function name and paren.';
var UNEXPECTED_SPACE = 'Unexpected space between function name and paren.';
var UNEXPECTED_NEWLINE = 'Unexpected newline between function name and paren.';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        var options = {
            expectSpace: false,
            spacePattern: /\s/
        };
        var userOptions = this.getOptions().ruleArguments;
        if (userOptions[0] === ALWAYS) {
            options.expectSpace = true;
            if (userOptions[1] !== undefined && userOptions[1].allowNewlines) {
                options.spacePattern = /[ \t\r\n\u2028\u2029]/;
            }
            else {
                options.spacePattern = /[ \t]/;
            }
        }
        var walker = new RuleWalker(sourceFile, RULE_NAME, options);
        return this.applyWithWalker(walker);
    };
    Rule.metadata = {
        ruleName: RULE_NAME,
        hasFix: true,
        description: 'require or disallow spacing between function identifiers and their invocations',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      This rule will enforce consistency of spacing in function calls,\n      by disallowing or requiring one or more spaces before the open paren.\n      "], ["\n      This rule will enforce consistency of spacing in function calls,\n      by disallowing or requiring one or more spaces before the open paren.\n      "]))),
        optionsDescription: Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      This rule has a string option:\n\n      * `\"never\"` (default) disallows space between the function name and the opening parenthesis.\n      * `\"always\"` requires space between the function name and the opening parenthesis.\n\n      Further, in `\"always\"` mode, a second object option is available that contains a single boolean `allowNewlines` property.\n      "], ["\n      This rule has a string option:\n\n      * \\`\"never\"\\` (default) disallows space between the function name and the opening parenthesis.\n      * \\`\"always\"\\` requires space between the function name and the opening parenthesis.\n\n      Further, in \\`\"always\"\\` mode, a second object option is available that contains a single boolean \\`allowNewlines\\` property.\n      "]))),
        options: {
            type: 'array',
            items: [
                {
                    enum: ['always', 'never']
                },
                {
                    type: 'object',
                    properties: {
                        allowNewlines: {
                            type: 'boolean'
                        }
                    },
                    additionalProperties: false
                }
            ],
            minItems: 0,
            maxItems: 2
        },
        optionExamples: [
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true]\n        "], ["\n        \"", "\": [true]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\"]\n        "], ["\n        \"", "\": [true, \"always\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\", { allowNewlines: true }]\n        "], ["\n        \"", "\": [true, \"always\", { allowNewlines: true }]\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'style'
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var RuleWalker = (function (_super) {
    tslib_1.__extends(RuleWalker, _super);
    function RuleWalker(sourceFile, ruleName, options) {
        var _this = _super.call(this, sourceFile, ruleName, options) || this;
        _this.sourceText = sourceFile.getFullText();
        return _this;
    }
    RuleWalker.prototype.walk = function (sourceFile) {
        var _this = this;
        var cb = function (node) {
            if (node.kind === ts.SyntaxKind.NewExpression) {
                _this.visitNewExpression(node);
            }
            else if (node.kind === ts.SyntaxKind.CallExpression) {
                _this.visitCallExpression(node);
            }
            else if (node.kind >= ts.SyntaxKind.FirstTypeNode && node.kind <= ts.SyntaxKind.LastTypeNode) {
                return;
            }
            return ts.forEachChild(node, cb);
        };
        return ts.forEachChild(sourceFile, cb);
    };
    RuleWalker.prototype.visitNewExpression = function (node) {
        this.checkWhitespaceAfterExpression(node.expression, node.typeArguments, node.arguments);
    };
    RuleWalker.prototype.visitCallExpression = function (node) {
        this.checkWhitespaceAfterExpression(node.expression, node.typeArguments, node.arguments);
    };
    RuleWalker.prototype.checkWhitespaceAfterExpression = function (expression, typeArguments, funcArguments) {
        if (funcArguments !== undefined) {
            var start = void 0;
            if (typeArguments !== undefined) {
                start = typeArguments.end + 1;
            }
            else {
                start = expression.getEnd();
            }
            this.checkWhitespaceBetween(start, funcArguments.pos - 1);
        }
    };
    RuleWalker.prototype.checkWhitespaceBetween = function (start, end) {
        var whitespace = this.sourceText.substring(start, end);
        if (this.options.spacePattern.test(whitespace)) {
            if (!this.options.expectSpace) {
                var fix = Lint.Replacement.deleteText(start, whitespace.length);
                var failureMessage = this.failureMessageForUnexpectedWhitespace(whitespace);
                this.addFailureAt(start, whitespace.length, failureMessage, fix);
            }
        }
        else if (this.options.expectSpace) {
            var fix = Lint.Replacement.appendText(start, ' ');
            this.addFailureAt(start, 1, MISSING_SPACE, fix);
        }
    };
    RuleWalker.prototype.failureMessageForUnexpectedWhitespace = function (whitespace) {
        if (/[\r\n]/.test(whitespace)) {
            return UNEXPECTED_NEWLINE;
        }
        else {
            return UNEXPECTED_SPACE;
        }
    };
    return RuleWalker;
}(Lint.AbstractWalker));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5;
