// tslint-eslint-rules/dist/rules/terNoSelfCompareRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var RULE_NAME = 'ter-no-self-compare';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new NoSelfCompareWalker(sourceFile, this.ruleName, new Set(this.ruleArguments.map(String))));
    };
    Rule.FAILURE_STRING = 'Comparing to itself is potentially pointless.';
    Rule.metadata = {
        ruleName: RULE_NAME,
        hasFix: false,
        description: 'disallow comparisons where both sides are exactly the same',
        rationale: 'Comparing a variable against itself is usually an error, ' +
            'either a typo or refactoring error. It is confusing to the reader ' +
            'and may potentially introduce a runtime error.',
        optionsDescription: '',
        options: {},
        optionExamples: [
            Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n        \"", "\": true\n        "], ["\n        \"", "\": true\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'maintainability'
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NoSelfCompareWalker = (function (_super) {
    tslib_1.__extends(NoSelfCompareWalker, _super);
    function NoSelfCompareWalker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NoSelfCompareWalker.prototype.isComparisonOperator = function (node) {
        var operators = new Set([
            ts.SyntaxKind.EqualsEqualsEqualsToken,
            ts.SyntaxKind.EqualsEqualsToken,
            ts.SyntaxKind.ExclamationEqualsEqualsToken,
            ts.SyntaxKind.ExclamationEqualsToken,
            ts.SyntaxKind.GreaterThanToken,
            ts.SyntaxKind.LessThanToken,
            ts.SyntaxKind.GreaterThanEqualsToken,
            ts.SyntaxKind.LessThanEqualsToken
        ]);
        return operators.has(node.operatorToken.kind);
    };
    NoSelfCompareWalker.prototype.hasSameToken = function (left, right) {
        return left.kind === right.kind && left.getText() === right.getText();
    };
    NoSelfCompareWalker.prototype.walk = function (sourceFile) {
        var _this = this;
        var cb = function (node) {
            if (ts.isBinaryExpression(node)) {
                var nodeExpr = node;
                if (_this.isComparisonOperator(nodeExpr) &&
                    _this.hasSameToken(nodeExpr.left, nodeExpr.right)) {
                    _this.addFailureAt(nodeExpr.operatorToken.getStart(), nodeExpr.operatorToken.getWidth(), Rule.FAILURE_STRING);
                }
            }
            else {
                return ts.forEachChild(node, cb);
            }
        };
        return ts.forEachChild(sourceFile, cb);
    };
    return NoSelfCompareWalker;
}(Lint.AbstractWalker));
var templateObject_1;
