// tslint-eslint-rules/dist/rules/terNoMixedSpacesAndTabsRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var tsutils_1 = require("tsutils");
var RULE_NAME = 'ter-no-mixed-spaces-and-tabs';
var OPTION_USE_TABS = 'tabs';
var OPTION_USE_SPACES = 'spaces';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.FAILURE_STRING = function (expected, mixed) {
        if (!mixed) {
            return expected + " indentation expected";
        }
        return "indentation has mixed tabs and spaces";
    };
    Rule.prototype.formatOptions = function (ruleArguments) {
        var tabs = undefined;
        var smartTabs = false;
        var options = ruleArguments[0];
        if (options !== undefined) {
            tabs = options.type === OPTION_USE_TABS ? true : options.type === OPTION_USE_SPACES ? false : undefined;
            smartTabs = options.smartTabs;
        }
        return {
            tabs: tabs,
            smartTabs: smartTabs
        };
    };
    Rule.prototype.apply = function (sourceFile) {
        var options = this.formatOptions(this.ruleArguments);
        return this.applyWithFunction(sourceFile, walk, options);
    };
    Rule.metadata = {
        ruleName: RULE_NAME,
        description: 'Enforces indentation with unmixed tabs or spaces.',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      Using only one of tabs or spaces for indentation leads to more consistent editor behavior,\n      cleaner diffs in version control, and easier programmatic manipulation."], ["\n      Using only one of tabs or spaces for indentation leads to more consistent editor behavior,\n      cleaner diffs in version control, and easier programmatic manipulation."]))),
        optionsDescription: Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      This rule takes an object argument with an optional `type` property which can be set to:\n\n      * `", "` enforces consistent spaces for indentation.\n      * `", "` enforces consistent tabs for indentation.\n\n      If the above is not provided, the rule will enforce either all tabs or all spaces on each\n      line, although different lines may differ between tabs and spaces.\n\n      Optionally, a `smartTabs` boolean property can be specified.  If set to true, smart tabs\n      allow mixing tabs and spaces if tabs are used for indentation and spaces for alignment, eg.\n\n          function main() {\n          // --->const a = 1,\n          // --->......b = 2;\n\n              const a = 1,\n                    b = 2;\n          }\n      "], ["\n      This rule takes an object argument with an optional \\`type\\` property which can be set to:\n\n      * \\`", "\\` enforces consistent spaces for indentation.\n      * \\`", "\\` enforces consistent tabs for indentation.\n\n      If the above is not provided, the rule will enforce either all tabs or all spaces on each\n      line, although different lines may differ between tabs and spaces.\n\n      Optionally, a \\`smartTabs\\` boolean property can be specified.  If set to true, smart tabs\n      allow mixing tabs and spaces if tabs are used for indentation and spaces for alignment, eg.\n\n          function main() {\n          // --->const a = 1,\n          // --->......b = 2;\n\n              const a = 1,\n                    b = 2;\n          }\n      "])), OPTION_USE_SPACES, OPTION_USE_TABS),
        options: {
            type: 'array',
            items: [
                {
                    type: 'object',
                    properties: {
                        type: {
                            type: 'string',
                            enum: [OPTION_USE_TABS, OPTION_USE_SPACES]
                        },
                        smartTabs: {
                            type: 'boolean'
                        }
                    },
                    additionalProperties: false
                }
            ],
            minLength: 0,
            maxLength: 1
        },
        optionExamples: [
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n      \"", "\": { \"type\": \"", "\" } ]\n      "], ["\n      \"", "\": { \"type\": \"", "\" } ]\n      "])), RULE_NAME, OPTION_USE_TABS),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n      \"", "\": { \"type\": \"", "\" } ]\n      "], ["\n      \"", "\": { \"type\": \"", "\" } ]\n      "])), RULE_NAME, OPTION_USE_SPACES),
            Lint.Utils.dedent(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n      \"", "\": { \"smartTabs\": true } ]\n      "], ["\n      \"", "\": { \"smartTabs\": true } ]\n      "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_6 || (templateObject_6 = tslib_1.__makeTemplateObject(["\n      \"", "\": { \"type\": \"", "\", \"smartTabs\": true } ]\n      "], ["\n      \"", "\": { \"type\": \"", "\", \"smartTabs\": true } ]\n      "])), RULE_NAME, OPTION_USE_TABS)
        ],
        type: 'maintainability',
        typescriptOnly: false
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
function walk(ctx) {
    var sourceFile = ctx.sourceFile, _a = ctx.options, tabs = _a.tabs, smartTabs = _a.smartTabs;
    var regExp;
    if (tabs === true) {
        regExp = new RegExp(" " + (smartTabs ? '\\t' : ''));
    }
    else if (tabs === false) {
        regExp = new RegExp("\\t");
    }
    else {
        regExp = new RegExp((smartTabs ? '' : '\\t |') + " \\t");
    }
    var failure = Rule.FAILURE_STRING(tabs ? 'tab' : 'space', typeof tabs === 'undefined');
    for (var _i = 0, _b = tsutils_1.getLineRanges(sourceFile); _i < _b.length; _i++) {
        var _c = _b[_i], pos = _c.pos, contentLength = _c.contentLength;
        if (contentLength === 0) {
            continue;
        }
        var line = sourceFile.text.substr(pos, contentLength);
        var indentEnd = line.search(/\S/);
        if (indentEnd === 0) {
            continue;
        }
        if (indentEnd === -1) {
            indentEnd = contentLength;
        }
        var indentSpace = line.slice(0, indentEnd);
        if (!regExp.test(indentSpace)) {
            continue;
        }
        var token = tsutils_1.getTokenAtPosition(sourceFile, pos);
        if (token.kind !== ts.SyntaxKind.JsxText &&
            (pos >= token.getStart(sourceFile) || tsutils_1.isPositionInComment(sourceFile, pos, token))) {
            continue;
        }
        ctx.addFailureAt(pos, indentEnd, failure);
    }
}
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
