// tslint-eslint-rules/dist/rules/terNewlineAfterVarRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var RULE_NAME = 'ter-newline-after-var';
var EXPECTED_BLANK_LINE_MESSAGE = 'Expected blank line after variable declarations.';
var UNEXPECTED_BLANK_LINE_MESSAGE = 'Unexpected blank line after variable declarations.';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.formatOptions = function (_a) {
        var alwaysOrNever = _a[0];
        return {
            always: alwaysOrNever !== 'never'
        };
    };
    Rule.prototype.apply = function (sourceFile) {
        var opt = this.formatOptions(this.ruleArguments);
        var walker = new RuleWalker(sourceFile, this.ruleName, opt);
        return this.applyWithWalker(walker);
    };
    Rule.metadata = {
        ruleName: RULE_NAME,
        hasFix: true,
        description: 'require or disallow an empty line after variable declarations',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      This rule enforces a coding style where empty lines are required or disallowed after `var`, `let`, or `const`\n      statements to achieve a consistent coding style across the project.\n      "], ["\n      This rule enforces a coding style where empty lines are required or disallowed after \\`var\\`, \\`let\\`, or \\`const\\`\n      statements to achieve a consistent coding style across the project.\n      "]))),
        optionsDescription: Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      This rule has a string option:\n\n      * `\"always\"` (default) requires an empty line after `var`, `let`, or `const`.\n        Comments on a line directly after var statements are treated like additional var statements.\n      * `\"never\"` disallows empty lines after `var`, `let`, or `const`.\n      "], ["\n      This rule has a string option:\n\n      * \\`\"always\"\\` (default) requires an empty line after \\`var\\`, \\`let\\`, or \\`const\\`.\n        Comments on a line directly after var statements are treated like additional var statements.\n      * \\`\"never\"\\` disallows empty lines after \\`var\\`, \\`let\\`, or \\`const\\`.\n      "]))),
        options: {
            type: 'array',
            items: [{
                    enum: ['always', 'never']
                }],
            maxLength: 1
        },
        optionExamples: [
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true]\n        "], ["\n        \"", "\": [true]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\"]\n        "], ["\n        \"", "\": [true, \"always\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"never\"]\n        "], ["\n        \"", "\": [true, \"never\"]\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'style'
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var RuleWalker = (function (_super) {
    tslib_1.__extends(RuleWalker, _super);
    function RuleWalker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RuleWalker.prototype.walk = function (sourceFile) {
        var _this = this;
        this.sourceFileText = sourceFile.getFullText();
        var onNode = function (node) {
            var _a = _this, lastVariableStatementNode = _a.lastVariableStatementNode, sourceFileText = _a.sourceFileText;
            if (node.kind === ts.SyntaxKind.VariableStatement) {
                _this.lastVariableStatementNode = node;
                return;
            }
            if (node.kind === ts.SyntaxKind.EndOfFileToken) {
                _this.lastVariableStatementNode = undefined;
                return;
            }
            if (lastVariableStatementNode) {
                var unexpectedLineFixes = [];
                var expectedLineFixes = [];
                var isNewLineRequired = _this.options.always;
                var expectedLinePos = lastVariableStatementNode.end;
                var newLinesCount = 0;
                for (var i = lastVariableStatementNode.end; i < node.end; i++) {
                    var code = sourceFileText.charCodeAt(i);
                    if (code === 10) {
                        newLinesCount++;
                        if (!isNewLineRequired && newLinesCount > 1) {
                            unexpectedLineFixes.push(Lint.Replacement.deleteText(i, 1));
                        }
                    }
                    else if (code !== 9 && code !== 13 && code !== 32) {
                        var leadingComments = ts.getLeadingCommentRanges("\n" + sourceFileText.slice(i), 0);
                        var lastLeadingComment = leadingComments && leadingComments.pop();
                        if (lastLeadingComment && (!isNewLineRequired || (isNewLineRequired && newLinesCount < 2))) {
                            newLinesCount = 0;
                            expectedLinePos = i - 1 + lastLeadingComment.end;
                            i = expectedLinePos - 1;
                        }
                        else {
                            if (isNewLineRequired && newLinesCount < 2) {
                                expectedLineFixes.push(Lint.Replacement.appendText(expectedLinePos, '\n'));
                            }
                            break;
                        }
                    }
                }
                if (isNewLineRequired && expectedLineFixes[0]) {
                    _this.addFailureAt(lastVariableStatementNode.getStart(), 1, EXPECTED_BLANK_LINE_MESSAGE, expectedLineFixes);
                }
                else if (unexpectedLineFixes[0]) {
                    _this.addFailureAt(lastVariableStatementNode.getStart(), 1, UNEXPECTED_BLANK_LINE_MESSAGE, unexpectedLineFixes);
                }
                _this.lastVariableStatementNode = undefined;
            }
            return ts.forEachChild(node, onNode);
        };
        return ts.forEachChild(sourceFile, onNode);
    };
    return RuleWalker;
}(Lint.AbstractWalker));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5;
