// tslint-eslint-rules/dist/rules/terPaddedBlocksRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var tsutils_1 = require("tsutils");
var RULE_NAME = 'ter-padded-blocks';
var OPTION_ALWAYS = 'always';
var Rule = (function (_super) {
    tslib_1.__extends(Rule, _super);
    function Rule() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Rule.prototype.formatOptions = function (ruleArguments) {
        var config = ruleArguments[0] || OPTION_ALWAYS;
        if (typeof (config) === 'string') {
            var always = config === OPTION_ALWAYS;
            return {
                blocks: always,
                classes: always,
                switches: always
            };
        }
        return {
            blocks: config['blocks'] && config['blocks'] === OPTION_ALWAYS,
            classes: config['classes'] && config['classes'] === OPTION_ALWAYS,
            switches: config['switches'] && config['switches'] === OPTION_ALWAYS
        };
    };
    Rule.prototype.apply = function (sourceFile) {
        var opt = this.formatOptions(this.ruleArguments);
        var walker = new RuleWalker(sourceFile, this.ruleName, opt);
        return this.applyWithWalker(walker);
    };
    Rule.metadata = {
        ruleName: RULE_NAME,
        hasFix: false,
        description: 'enforces consistent empty line padding within blocks',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      Some style guides require block statements to start and end with blank\n      lines. The goal is to improve readability by visually separating the\n      block content and the surrounding code.\n      "], ["\n      Some style guides require block statements to start and end with blank\n      lines. The goal is to improve readability by visually separating the\n      block content and the surrounding code.\n      "]))),
        optionsDescription: 'This rule has one option, which can be a string option or an object option',
        options: {
            type: 'array',
            items: [
                {
                    enum: ['always', 'never']
                },
                {
                    type: 'object',
                    properties: {
                        blocks: {
                            enum: ['always', 'never']
                        },
                        classes: {
                            enum: ['always', 'never']
                        },
                        switches: {
                            enum: ['always', 'never']
                        }
                    },
                    additionalProperties: false
                }
            ],
            maxLength: 1
        },
        optionExamples: [
            Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true]\n        "], ["\n        \"", "\": [true]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\"]\n        "], ["\n        \"", "\": [true, \"always\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"never\"]\n        "], ["\n        \"", "\": [true, \"never\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, { \"blocks\": \"always\" }]\n        "], ["\n        \"", "\": [true, { \"blocks\": \"always\" }]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_6 || (templateObject_6 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, { \"blocks\": \"never\" }]\n        "], ["\n        \"", "\": [true, { \"blocks\": \"never\" }]\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'style'
    };
    Rule.FAILURE_STRING = {
        always: 'Block must be padded by blank lines.',
        never: 'Block must not be padded by blank lines.'
    };
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var RuleWalker = (function (_super) {
    tslib_1.__extends(RuleWalker, _super);
    function RuleWalker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RuleWalker.prototype.walk = function (sourceFile) {
        var _this = this;
        sourceFile.forEachChild(function (node) { return _this.processNode(node); });
    };
    RuleWalker.prototype.processNode = function (node) {
        var _this = this;
        switch (node.kind) {
            case ts.SyntaxKind.Block:
            case ts.SyntaxKind.ClassDeclaration:
            case ts.SyntaxKind.CaseBlock:
                this.checkPadding(node);
        }
        node.forEachChild(function (child) { return _this.processNode(child); });
    };
    RuleWalker.prototype.getParts = function (node) {
        var openBrace, body, closeBrace;
        node.getChildren().forEach(function (child) {
            if (child.kind === ts.SyntaxKind.OpenBraceToken) {
                openBrace = child;
            }
            else if (child.kind === ts.SyntaxKind.SyntaxList) {
                body = child;
            }
            else if (child.kind === ts.SyntaxKind.CloseBraceToken) {
                closeBrace = child;
            }
        });
        return {
            openBrace: openBrace,
            body: body,
            closeBrace: closeBrace
        };
    };
    RuleWalker.prototype.getPositions = function (node) {
        var _a = this.getParts(node), openBrace = _a.openBrace, body = _a.body, closeBrace = _a.closeBrace;
        var firstChild = body.getChildAt(0);
        var lastChild = body.getChildAt(body.getChildCount() - 1);
        var positions = {
            openPosition: openBrace.getStart(this.sourceFile),
            openLine: this.getLine(openBrace.getStart(this.sourceFile)),
            closePosition: closeBrace.getEnd(),
            closeLine: this.getLine(closeBrace.getEnd())
        };
        if (firstChild) {
            positions.firstChildPosition = firstChild.getStart(this.sourceFile);
            positions.firstChildLine = this.getLine(positions.firstChildPosition);
        }
        if (lastChild) {
            positions.lastChildPosition = lastChild.getEnd();
            positions.lastChildLine = this.getLine(positions.lastChildPosition);
        }
        return positions;
    };
    RuleWalker.prototype.checkPadding = function (node) {
        var _this = this;
        var paddingAllowed = this.options.blocks;
        if (node.kind === ts.SyntaxKind.ClassDeclaration) {
            paddingAllowed = this.options.classes;
        }
        else if (node.parent && node.parent.kind === ts.SyntaxKind.SwitchStatement) {
            paddingAllowed = this.options.switches;
        }
        if (paddingAllowed === undefined) {
            return;
        }
        var positions = this.getPositions(node);
        var openBraceReplacement = {
            from: positions.openPosition + 1,
            to: positions.firstChildPosition || positions.closePosition
        };
        var closeBraceReplacement = {
            from: positions.lastChildPosition || positions.openPosition,
            to: positions.closePosition - 1
        };
        var comments = [];
        tsutils_1.forEachComment(node, function (_fullText, comment) {
            if (comment.pos > positions.openPosition && comment.pos < positions.closePosition) {
                var commentLineEnd = _this.getLine(comment.end);
                if (commentLineEnd > positions.openLine) {
                    comments.push(comment);
                }
                else if (commentLineEnd === positions.openLine) {
                    openBraceReplacement.from = comment.end;
                }
            }
        });
        if (comments.length > 0) {
            var firstCommentLine = this.getLine(comments[0].pos);
            var lastCommentLine = this.getLine(comments[comments.length - 1].end);
            if (!positions.firstChildLine || firstCommentLine < positions.firstChildLine) {
                positions.firstChildLine = firstCommentLine;
                positions.firstChildPosition = comments[0].pos;
                openBraceReplacement.to = positions.firstChildPosition;
            }
            if (!positions.lastChildLine || lastCommentLine >= positions.lastChildLine) {
                positions.lastChildLine = lastCommentLine;
                positions.lastChildPosition = comments[comments.length - 1].end;
                closeBraceReplacement.from = positions.lastChildPosition;
            }
        }
        if (this.getLine(openBraceReplacement.from) !== this.getLine(openBraceReplacement.to)) {
            openBraceReplacement.to = this.getPosition(this.getLine(openBraceReplacement.to));
        }
        if (this.getLine(closeBraceReplacement.from) !== this.getLine(closeBraceReplacement.to)) {
            closeBraceReplacement.to = this.getPosition(this.getLine(closeBraceReplacement.to));
        }
        if (positions.firstChildLine === undefined && positions.lastChildLine === undefined) {
            return;
        }
        var openPadded = false;
        if (positions.firstChildLine !== undefined) {
            openPadded = positions.firstChildLine - positions.openLine > 1;
        }
        else {
            openPadded = positions.closeLine - positions.openLine > 1;
        }
        var closePadded = false;
        if (positions.lastChildLine !== undefined) {
            closePadded = positions.closeLine - positions.lastChildLine > 1;
        }
        else {
            closePadded = positions.closeLine - positions.openLine > 1;
        }
        if (paddingAllowed ? !openPadded : openPadded) {
            var openFix = Lint.Replacement.replaceFromTo(openBraceReplacement.from, openBraceReplacement.to, paddingAllowed ? '\n\n' : '\n');
            this.addFailure(positions.openPosition, positions.openPosition + 1, paddingAllowed ? Rule.FAILURE_STRING.always : Rule.FAILURE_STRING.never, openFix);
        }
        if (paddingAllowed ? !closePadded : closePadded) {
            var closeFix = Lint.Replacement.replaceFromTo(closeBraceReplacement.from, closeBraceReplacement.to, paddingAllowed ? '\n\n' : '\n');
            this.addFailure(positions.closePosition - 1, positions.closePosition, paddingAllowed ? Rule.FAILURE_STRING.always : Rule.FAILURE_STRING.never, closeFix);
        }
    };
    RuleWalker.prototype.getLine = function (pos) {
        return this.sourceFile.getLineAndCharacterOfPosition(pos).line;
    };
    RuleWalker.prototype.getPosition = function (line) {
        return this.sourceFile.getPositionOfLineAndCharacter(line, 0);
    };
    return RuleWalker;
}(Lint.AbstractWalker));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
